<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>API &mdash; SCENIC+ 0.1.dev132+g26f1691 documentation</title><link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "tex2jax_ignore|mathjax_ignore|document", "processClass": "tex2jax_process|mathjax_process|math|output_area"}})</script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorials" href="tutorials.html" />
    <link rel="prev" title="Install" href="install.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> SCENIC+
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Install</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-scenicplus.scenicplus_class">SCENIC+ object</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-scenicplus.preprocessing.filtering">Preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="#scenic-semi-automated-workflow-using-wrapper-functions">SCENIC+ semi-automated workflow using wrapper functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-scenicplus.wrappers.run_pycistarget">pycistarget wrapper</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-scenicplus.wrappers.run_scenicplus">SCENIC+ wrapper</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#tools-for-non-automated-workflow">Tools for non-automated workflow</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-scenicplus.cistromes">Cistromes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-scenicplus.enhancer_to_gene">Enhancer-to-gene linking</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-scenicplus.TF_to_gene">TF-to-gene linking</a></li>
<li class="toctree-l3"><a class="reference internal" href="#egrn-building">eGRN building</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#module-scenicplus.grn_builder.modules">eRegulon Class</a></li>
<li class="toctree-l4"><a class="reference internal" href="#gsea-based-approach">GSEA based approach</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#downstream-analysis-export-and-plotting">Downstream analysis, export and plotting</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-scenicplus.diff_features">Marker genes and regions</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-scenicplus.eregulon_enrichment">eRegulon enrichment in cells</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-scenicplus.dimensionality_reduction">dimensionality reduction</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-scenicplus.RSS">eRegulon specificity score (eRSS)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-scenicplus.networks">Network</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-scenicplus.plotting.correlation_plot">Correlation plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-scenicplus.plotting.coverageplot">Coverage plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-scenicplus.plotting.dotplot">dotplot</a></li>
<li class="toctree-l3"><a class="reference internal" href="#export">Export</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#module-scenicplus.loom">Export to loom</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-scenicplus.hdf5">Export to hdf5</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">SCENIC+</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>API</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="api">
<span id="id1"></span><h1>API<a class="headerlink" href="#api" title="Permalink to this headline"></a></h1>
<div class="section" id="module-scenicplus.scenicplus_class">
<span id="scenic-object"></span><h2>SCENIC+ object<a class="headerlink" href="#module-scenicplus.scenicplus_class" title="Permalink to this headline"></a></h2>
<p>Combine single-cell expression data and single-cell accessibility into a single SCENIC+ object.</p>
<p>This object will be used for downstream analysis, including: region-to-gene and TF-to-gene linking, enhancer-driven-GRN building and further downstream analysis.</p>
<p>This object can be generated from both single-cell multi-omics data (i.e. gene expression and chromatin accessibility from the same cell),
and seperate single-cell chromatin accessbility data and single-cell gene expression data from the same or similar sample.</p>
<p>In the second case both data modalities should have a common cell metadata field with values linking both modalities (e.g. common celltype annotation).</p>
<dl class="py class">
<dt id="scenicplus.scenicplus_class.SCENICPLUS">
<em class="property">class </em><code class="sig-prename descclassname">scenicplus.scenicplus_class.</code><code class="sig-name descname">SCENICPLUS</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">X_ACC</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>scipy.sparse.base.spmatrix<span class="p">, </span>numpy.ndarray<span class="p">, </span>pandas.core.frame.DataFrame<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">X_EXP</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>scipy.sparse.base.spmatrix<span class="p">, </span>numpy.ndarray<span class="p">, </span>pandas.core.frame.DataFrame<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">metadata_regions</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">metadata_genes</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">metadata_cell</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">menr</span><span class="p">:</span> <span class="n">Mapping<span class="p">[</span>str<span class="p">, </span>Mapping<span class="p">[</span>str<span class="p">, </span>Any<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">dr_cell</span><span class="p">:</span> <span class="n">Mapping<span class="p">[</span>str<span class="p">, </span>numpy.iterable<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">dr_region</span><span class="p">:</span> <span class="n">Mapping<span class="p">[</span>str<span class="p">, </span>numpy.iterable<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">uns</span><span class="p">:</span> <span class="n">Mapping<span class="p">[</span>str<span class="p">, </span>Any<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{}</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/scenicplus_class.html#SCENICPLUS"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.scenicplus_class.SCENICPLUS" title="Permalink to this definition"></a></dt>
<dd><p>An object containing: gene expression, chromatin accessbility and motif enrichment data.</p>
<p><code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code> stores the data matrices <code class="xref py py-attr docutils literal notranslate"><span class="pre">X_ACC</span></code> (chromatin accessbility) and
<code class="xref py py-attr docutils literal notranslate"><span class="pre">X_EXP</span></code> (gene expression) together with region annotation <code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_regions</span></code>, gene annotation <code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_genes</span></code>,
cell annotation <code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_cell</span></code> and motif enrichment data <code class="xref py py-attr docutils literal notranslate"><span class="pre">menr</span></code>.</p>
<p>Use create_SCENICPLUS_object to generate instances of this class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>X_ACC: sparse.spmatrix, np.ndarray, pd.DataFrame</strong></dt><dd><p>A #regions x #cells data matrix</p>
</dd>
<dt><strong>X_EXP: sparse.spmatrix, np.ndarray, pd.DataFrame</strong></dt><dd><p>A #cells x #genes data matrix</p>
</dd>
<dt><strong>metadata_regions: pd.DataFrame</strong></dt><dd><p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> containing region metadata annotation of length #regions</p>
</dd>
<dt><strong>metadata_genes: pd.DataFrame</strong></dt><dd><p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> containing gene metadata annotation of length #genes</p>
</dd>
<dt><strong>metadata_cell: pd.DataFrame</strong></dt><dd><p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> containing cell metadata annotation of lenght #cells</p>
</dd>
<dt><strong>menr: dict</strong></dt><dd><p>A Dict containing motif enrichment results for topics of differentially accessbile regions (DARs), generate by pycistarget.
Should take the form {‘region_set_name1’: {region_set1: result, ‘region_set2’: result}, ‘region_set_name2’: {‘region_set1’: result, ‘region_set2’: result},
‘topic’: {‘topic1’: result, ‘topic2’: result}}
region set names, which aren’t topics, should be columns in the <code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_cell</span></code> dataframe</p>
</dd>
<dt><strong>dr_cell: dict</strong></dt><dd><p>A Dict containing dimmensional reduction coordinates of cells.</p>
</dd>
<dt><strong>dr_region: dct</strong></dt><dd><p>A Dict containing dimmensional reduction coordinates of regions.</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#scenicplus.scenicplus_class.create_SCENICPLUS_object" title="scenicplus.scenicplus_class.create_SCENICPLUS_object"><code class="xref py py-obj docutils literal notranslate"><span class="pre">scenicplus.scenicplus_class.create_SCENICPLUS_object</span></code></a></dt><dd></dd>
</dl>
</div>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>n_cells: int</strong></dt><dd><p>Returns number of cells.</p>
</dd>
<dt><strong>n_genes: int</strong></dt><dd><p>Returns number of genes.</p>
</dd>
<dt><strong>n_regions: int</strong></dt><dd><p>Returns number of regions.</p>
</dd>
<dt><strong>cell_names: List[str]</strong></dt><dd><p>Returns cell names</p>
</dd>
<dt><strong>gene_names: List[str]</strong></dt><dd><p>Returns gene names</p>
</dd>
<dt><strong>region_names: List[str]</strong></dt><dd><p>Returns region names</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS.add_cell_data" title="scenicplus.scenicplus_class.SCENICPLUS.add_cell_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_cell_data</span></code></a>(cell_data)</p></td>
<td><p>Add cell metadata</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS.add_gene_data" title="scenicplus.scenicplus_class.SCENICPLUS.add_gene_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_gene_data</span></code></a>(gene_data)</p></td>
<td><p>Add gene metadata</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS.add_region_data" title="scenicplus.scenicplus_class.SCENICPLUS.add_region_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_region_data</span></code></a>(region_data)</p></td>
<td><p>Add region metadata</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS.subset" title="scenicplus.scenicplus_class.SCENICPLUS.subset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">subset</span></code></a>([cells, regions, genes, return_copy])</p></td>
<td><p>Subset object</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS.to_df" title="scenicplus.scenicplus_class.SCENICPLUS.to_df"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_df</span></code></a>(layer)</p></td>
<td><p>Generate a <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt id="scenicplus.scenicplus_class.SCENICPLUS.add_cell_data">
<code class="sig-name descname">add_cell_data</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cell_data</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/scenicplus_class.html#SCENICPLUS.add_cell_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.scenicplus_class.SCENICPLUS.add_cell_data" title="Permalink to this definition"></a></dt>
<dd><p>Add cell metadata</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>cell_data: pd.DataFrame</strong></dt><dd><p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> containing cell metdata indexed with cell barcodes.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="scenicplus.scenicplus_class.SCENICPLUS.add_gene_data">
<code class="sig-name descname">add_gene_data</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_data</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/scenicplus_class.html#SCENICPLUS.add_gene_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.scenicplus_class.SCENICPLUS.add_gene_data" title="Permalink to this definition"></a></dt>
<dd><p>Add gene metadata</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gene_data: pd.DataFrame</strong></dt><dd><p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> containing gene metadata indexed with gene names.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="scenicplus.scenicplus_class.SCENICPLUS.add_region_data">
<code class="sig-name descname">add_region_data</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">region_data</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/scenicplus_class.html#SCENICPLUS.add_region_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.scenicplus_class.SCENICPLUS.add_region_data" title="Permalink to this definition"></a></dt>
<dd><p>Add region metadata</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>region_data: pd.DataFrame</strong></dt><dd><p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> containing region metadata indexed with region names.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="scenicplus.scenicplus_class.SCENICPLUS.subset">
<code class="sig-name descname">subset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cells</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">regions</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">genes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">return_copy</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/scenicplus_class.html#SCENICPLUS.subset"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.scenicplus_class.SCENICPLUS.subset" title="Permalink to this definition"></a></dt>
<dd><p>Subset object</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>cells: List[str]</strong></dt><dd><p>A list of cells to keep
default: None</p>
</dd>
<dt><strong>regions: List[str]</strong></dt><dd><p>A list of regions to keep
default: None</p>
</dd>
<dt><strong>genes: List[str]</strong></dt><dd><p>A list of genes to keep
default:None</p>
</dd>
<dt><strong>return_copy: bool</strong></dt><dd><p>A boolean specifying wether to update the object (False) or return a copy (True)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="scenicplus.scenicplus_class.SCENICPLUS.to_df">
<code class="sig-name descname">to_df</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">layer</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="reference internal" href="_modules/scenicplus/scenicplus_class.html#SCENICPLUS.to_df"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.scenicplus_class.SCENICPLUS.to_df" title="Permalink to this definition"></a></dt>
<dd><p>Generate a <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>.</p>
<p>The data matrix <code class="xref py py-attr docutils literal notranslate"><span class="pre">X_EXP</span></code> or <code class="xref py py-attr docutils literal notranslate"><span class="pre">X_ACC</span></code> is returned as
<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>, with <code class="xref py py-attr docutils literal notranslate"><span class="pre">cell_names</span></code> as index,
and <code class="xref py py-attr docutils literal notranslate"><span class="pre">gene_names</span></code> or <code class="xref py py-attr docutils literal notranslate"><span class="pre">region_names</span></code> as columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>layer: str</strong></dt><dd><p>ACC to return accessibility data and EXP to return expression data.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="scenicplus.scenicplus_class.create_SCENICPLUS_object">
<code class="sig-prename descclassname">scenicplus.scenicplus_class.</code><code class="sig-name descname">create_SCENICPLUS_object</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">GEX_anndata</span><span class="p">:</span> <span class="n">anndata._core.anndata.AnnData</span></em>, <em class="sig-param"><span class="n">cisTopic_obj</span><span class="p">:</span> <span class="n">pycisTopic.cistopic_class.CistopicObject</span></em>, <em class="sig-param"><span class="n">menr</span><span class="p">:</span> <span class="n">Mapping<span class="p">[</span>str<span class="p">, </span>Mapping<span class="p">[</span>str<span class="p">, </span>Any<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">multi_ome_mode</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">nr_metacells</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>int<span class="p">, </span>Mapping<span class="p">[</span>str<span class="p">, </span>int<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">nr_cells_per_metacells</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>int<span class="p">, </span>Mapping<span class="p">[</span>str<span class="p">, </span>int<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">10</span></em>, <em class="sig-param"><span class="n">meta_cell_split</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'_'</span></em>, <em class="sig-param"><span class="n">key_to_group_by</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">imputed_acc_obj</span><span class="p">:</span> <span class="n">pycisTopic.diff_features.CistopicImputedFeatures</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">imputed_acc_kwargs</span><span class="p">:</span> <span class="n">Mapping<span class="p">[</span>str<span class="p">, </span>Any<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'scale_factor': 1000000}</span></em>, <em class="sig-param"><span class="n">normalize_imputed_acc</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">normalize_imputed_acc_kwargs</span><span class="p">:</span> <span class="n">Mapping<span class="p">[</span>str<span class="p">, </span>Any<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'scale_factor': 10000}</span></em>, <em class="sig-param"><span class="n">cell_metadata</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">region_metadata</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gene_metadata</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">bc_transform_func</span><span class="p">:</span> <span class="n">Callable</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ACC_prefix</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'ACC_'</span></em>, <em class="sig-param"><span class="n">GEX_prefix</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'GEX_'</span></em><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a><a class="reference internal" href="_modules/scenicplus/scenicplus_class.html#create_SCENICPLUS_object"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.scenicplus_class.create_SCENICPLUS_object" title="Permalink to this definition"></a></dt>
<dd><p>Function to create instances of <a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS"><code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>GEX_anndata: sc.AnnData</strong></dt><dd><p>An instance of <code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code> containing gene expression data and metadata.</p>
</dd>
<dt><strong>cisTopic_obj: CistopicObject</strong></dt><dd><p>An instance of <code class="xref py py-class docutils literal notranslate"><span class="pre">pycisTopic.cistopic_class.CistopicObject</span></code> containing chromatin accessibility data and metadata.</p>
</dd>
<dt><strong>menr: dict</strong></dt><dd><p>A dict mapping annotations to motif enrichment results</p>
</dd>
<dt><strong>multi_ome_mode: bool</strong></dt><dd><p>A boolean specifying wether data is multi-ome (i.e. combined scATAC-seq and scRNA-seq from the same cell) or not
default: True</p>
</dd>
<dt><strong>nr_metacells: int</strong></dt><dd><p>For non multi_ome_mode, use this number of meta cells to link scRNA-seq and scATAC-seq
If this is a single integer the same number of metacells will be used for all annotations.
This can also be a mapping between an annotation and the number of metacells per annotation.
default: None</p>
</dd>
<dt><strong>nr_cells_per_metacells: int</strong></dt><dd><p>For non multi_ome_mode, use this number of cells per metacell to link scRNA-seq and scATAC-seq.
If this is a single integer the same number of cells will be used for all annotations.
This can also be a mapping between an annotation and the number of cells per metacell per annotation.
default: 10</p>
</dd>
<dt><strong>meta_cell_split: str</strong></dt><dd><p>Character which is used as seperator in metacell names
default: ‘_’</p>
</dd>
<dt><strong>key_to_group_by: str</strong></dt><dd><p>For non multi_ome_mode, use this cell metadata key to generate metacells from scRNA-seq and scATAC-seq.
Key should be common in scRNA-seq and scATAC-seq side
default: None</p>
</dd>
<dt><strong>imputed_acc_obj: CistopicImputedFeatures</strong></dt><dd><p>An instance of <code class="xref py py-class docutils literal notranslate"><span class="pre">CistopicImputedFeatures</span></code> containing imputed chromatin accessibility.
default: None</p>
</dd>
<dt><strong>imputed_acc_kwargs: dict</strong></dt><dd><p>Dict with keyword arguments for imputed chromatin accessibility.
default: {‘scale_factor’: 10**6}</p>
</dd>
<dt><strong>normalize_imputed_acc: bool</strong></dt><dd><p>A boolean specifying wether or not to normalize imputed chromatin accessbility.
default: False</p>
</dd>
<dt><strong>normalize_imputed_acc_kwargs: dict</strong></dt><dd><p>Dict with keyword arguments for normalizing imputed accessibility.
default: {‘scale_factor’: 10 ** 4}</p>
</dd>
<dt><strong>cell_metadata: pd.DataFrame</strong></dt><dd><p>An instance of <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> containing extra cell metadata
default: None</p>
</dd>
<dt><strong>region_metadata: pd.DataFrame</strong></dt><dd><p>An instance of <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> containing extra region metadata
default: None</p>
</dd>
<dt><strong>gene_metadata: pd.DataFrame</strong></dt><dd><p>An instance of <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> containing extra gene metadata
default: None</p>
</dd>
<dt><strong>bc_transform_func: func</strong></dt><dd><p>A function used to transform gene expression barcode layout to chromatin accessbility layout.
default: None</p>
</dd>
<dt><strong>ACC_prefix: str</strong></dt><dd><p>String prefix to add to cell metadata coming from cisTopic_obj</p>
</dd>
<dt><strong>GEX_prefix: str</strong></dt><dd><p>String prefix to add to cell metadata coming from GEX_anndata</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">scplus_obj</span> <span class="o">=</span> <span class="n">create_SCENICPLUS_object</span><span class="p">(</span><span class="n">GEX_anndata</span> <span class="o">=</span> <span class="n">rna_anndata</span><span class="p">,</span> <span class="n">cisTopic_obj</span> <span class="o">=</span> <span class="n">cistopic_obj</span><span class="p">,</span> <span class="n">menr</span> <span class="o">=</span> <span class="n">motif_enrichment_dict</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="module-scenicplus.preprocessing.filtering">
<span id="preprocessing"></span><h2>Preprocessing<a class="headerlink" href="#module-scenicplus.preprocessing.filtering" title="Permalink to this headline"></a></h2>
<p>Filter outlier genes and regions.</p>
<dl class="py function">
<dt id="scenicplus.preprocessing.filtering.filter_genes">
<code class="sig-prename descclassname">scenicplus.preprocessing.filtering.</code><code class="sig-name descname">filter_genes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">SCENICPLUS_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">min_pct</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">0</span></em>, <em class="sig-param"><span class="n">max_pct</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">100</span></em>, <em class="sig-param"><span class="n">return_copy</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a><a class="reference internal" href="_modules/scenicplus/preprocessing/filtering.html#filter_genes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.preprocessing.filtering.filter_genes" title="Permalink to this definition"></a></dt>
<dd><p>Filter scenciplus object genes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>SCENICPLUS_obj</strong></dt><dd><p>An instance of :class: <cite>~scenicplus.scenicplus_class.SCENICPLUS</cite>.</p>
</dd>
<dt><strong>min_pct</strong></dt><dd><p>only keep genes which are expressed in at least <cite>min_pct</cite> of cells.
default: 0</p>
</dd>
<dt><strong>max_pct</strong></dt><dd><p>only keep genes which are expressed in maximal <cite>max_pct</cite> of cells.
default: 100</p>
</dd>
<dt><strong>return_copy</strong></dt><dd><p>If set to True a new SCENICPLUS object will be generated containing filtered data.
default: False</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.preprocessing.filtering.filter_regions">
<code class="sig-prename descclassname">scenicplus.preprocessing.filtering.</code><code class="sig-name descname">filter_regions</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">SCENICPLUS_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">min_pct</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">0</span></em>, <em class="sig-param"><span class="n">max_pct</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">100</span></em>, <em class="sig-param"><span class="n">return_copy</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a><a class="reference internal" href="_modules/scenicplus/preprocessing/filtering.html#filter_regions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.preprocessing.filtering.filter_regions" title="Permalink to this definition"></a></dt>
<dd><p>Filter scenciplus object regions</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>SCENICPLUS_obj</strong></dt><dd><p>An instance of :class: <cite>~scenicplus.scenicplus_class.SCENICPLUS</cite>.</p>
</dd>
<dt><strong>min_pct</strong></dt><dd><p>only keep regions which are accessible in at least <cite>min_pct</cite> of cells.
default: 0</p>
</dd>
<dt><strong>max_pct</strong></dt><dd><p>only keep regions which are accessible in maximal <cite>max_pct</cite> of cells.
default: 100</p>
</dd>
<dt><strong>return_copy</strong></dt><dd><p>If set to True a new SCENICPLUS object will be generated containing filtered data.
default: False</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="scenic-semi-automated-workflow-using-wrapper-functions">
<h2>SCENIC+ semi-automated workflow using wrapper functions<a class="headerlink" href="#scenic-semi-automated-workflow-using-wrapper-functions" title="Permalink to this headline"></a></h2>
<div class="section" id="module-scenicplus.wrappers.run_pycistarget">
<span id="pycistarget-wrapper"></span><h3>pycistarget wrapper<a class="headerlink" href="#module-scenicplus.wrappers.run_pycistarget" title="Permalink to this headline"></a></h3>
<p>Wrapper functions to run motif enrichment analysis using pycistarget</p>
<p>After sets of regions have been defined (e.g. topics or DARs). The complete pycistarget workflo can be run using a single function.</p>
<p>this function will run cistarget based and DEM based motif enrichment analysis with or without promoter regions.</p>
<dl class="py function">
<dt id="scenicplus.wrappers.run_pycistarget.run_pycistarget">
<code class="sig-prename descclassname">scenicplus.wrappers.run_pycistarget.</code><code class="sig-name descname">run_pycistarget</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">region_sets</span><span class="p">:</span> <span class="n">Mapping<span class="p">[</span>str<span class="p">, </span>pyranges.pyranges.PyRanges<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">species</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">save_path</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">ctx_db_path</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">dem_db_path</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">run_without_promoters</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">biomart_host</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'http://www.ensembl.org'</span></em>, <em class="sig-param"><span class="n">promoter_space</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">500</span></em>, <em class="sig-param"><span class="n">ctx_auc_threshold</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.005</span></em>, <em class="sig-param"><span class="n">ctx_nes_threshold</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">3.0</span></em>, <em class="sig-param"><span class="n">ctx_rank_threshold</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.05</span></em>, <em class="sig-param"><span class="n">dem_log2fc_thr</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">dem_motif_hit_thr</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">3.0</span></em>, <em class="sig-param"><span class="n">dem_max_bg_regions</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">500</span></em>, <em class="sig-param"><span class="n">annotation</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">['Direct_annot', 'Orthology_annot']</span></em>, <em class="sig-param"><span class="n">path_to_motif_annotations</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">annotation_version</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'v9'</span></em>, <em class="sig-param"><span class="n">n_cpu</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">1</span></em>, <em class="sig-param"><span class="n">_temp_dir</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">exclude_motifs</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">exclude_collection</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/wrappers/run_pycistarget.html#run_pycistarget"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.wrappers.run_pycistarget.run_pycistarget" title="Permalink to this definition"></a></dt>
<dd><p>Wrapper function for pycistarget</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region_sets</strong> (<em>Mapping</em><em>[</em><em>str</em><em>, </em><em>pr.PyRanges</em><em>]</em>) – A dictionary of PyRanges containing region coordinates for the region sets to be analyzed.</p></li>
<li><p><strong>species</strong> (<em>str</em>) – Species from which genomic coordinates come from, options are: homo_sapiens, mus_musculus and drosophila_melanogaster.</p></li>
<li><p><strong>save_path</strong> (<em>str</em>) – Directory in which to save outputs.</p></li>
<li><p><strong>ctx_db_path</strong> (<em>str = None</em>) – Path to cistarget database containing rankings of motif scores</p></li>
<li><p><strong>dem_db_path</strong> (<em>str = None</em>) – Path to dem database containing motif scores</p></li>
<li><p><strong>run_without_promoters</strong> (<em>bool = False</em>) – Boolean specifying wether the analysis should also be run without including promoter regions.</p></li>
<li><p><strong>biomart_host</strong> (str = ‘<a class="reference external" href="http://www.ensembl.org">http://www.ensembl.org</a>’) – url to biomart host, make sure this host matches your assembly</p></li>
<li><p><strong>promoter_space</strong> (<em>int = 500</em>) – integer defining space around the TSS to consider as promoter</p></li>
<li><p><strong>ctx_auc_threshold</strong> (<em>float = 0.005</em>) – The fraction of the ranked genome to take into account for the calculation of the Area Under the recovery Curve</p></li>
<li><p><strong>ctx_nes_threshold</strong> (<em>float = 3.0</em>) – The Normalized Enrichment Score (NES) threshold to select enriched features.</p></li>
<li><p><strong>ctx_rank_threshold</strong> (<em>float = 0.05</em>) – The total number of ranked genes to take into account when creating a recovery curve.</p></li>
<li><p><strong>dem_log2fc_thr</strong> (<em>float = 0.5</em>) – Log2 Fold-change threshold to consider a motif enriched.</p></li>
<li><p><strong>dem_motif_hit_thr</strong> (<em>float = 3.0</em>) – Minimul mean signal in the foreground to consider a motif enriched.</p></li>
<li><p><strong>dem_max_bg_regions</strong> (<em>int = 500</em>) – Maximum number of regions to use as background. When set to None, all regions are used</p></li>
<li><p><strong>annotation</strong> (<em>List</em><em>[</em><em>str</em><em>] </em><em>=</em><em> [</em><em>'Direct_annot'</em><em>, </em><em>'Orthology_annot'</em><em>]</em>) – Annotation to use for forming cistromes. It can be ‘Direct_annot’ (direct evidence that the motif is
linked to that TF), ‘Motif_similarity_annot’ (based on tomtom motif similarity), ‘Orthology_annot’
(based on orthology with a TF that is directly linked to that motif) or ‘Motif_similarity_and_Orthology_annot’.</p></li>
<li><p><strong>path_to_motif_annotations</strong> (<em>str = None</em>) – Path to motif annotations. If not provided, they will be downloaded from
<a class="reference external" href="https://resources.aertslab.org">https://resources.aertslab.org</a> based on the specie name provided (only possible for mus_musculus,
homo_sapiens and drosophila_melanogaster).</p></li>
<li><p><strong>annotation_version</strong> (<em>str = 'v9'</em>) – Motif collection version.</p></li>
<li><p><strong>n_cpu</strong> (<em>int = 1</em>) – Number of cores to use.</p></li>
<li><p><strong>_temp_dir</strong> (<em>str = None</em>) – temp_dir to use for ray.</p></li>
<li><p><strong>exclude_motifs</strong> (<em>str = None</em>) – Path to csv file containing motif to exclude from the analysis.</p></li>
<li><p><strong>exclude_collection</strong> (<em>List</em><em>[</em><em>str</em><em>] </em><em>= None</em>) – List of strings identifying which motif collections to exclude from analysis.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-scenicplus.wrappers.run_scenicplus">
<span id="scenic-wrapper"></span><h3>SCENIC+ wrapper<a class="headerlink" href="#module-scenicplus.wrappers.run_scenicplus" title="Permalink to this headline"></a></h3>
<p>Wrapper functions to run SCENIC+ analysis</p>
<p>After the SCENIC+ object has been generated the complete SCENIC+ workflow can be run with a single function.</p>
<p>Following operation will be done:</p>
<ol class="arabic simple">
<li><p>Defining a search space surounding each gene</p></li>
<li><p>Region to gene linking &amp; TF to gene linking</p></li>
<li><p>eGRN building</p></li>
<li><p>eGRN scoring (target gene and region based AUC and RSS)</p></li>
<li><p>Calculating TF-cistrome correlations</p></li>
<li><p>Dimensionality reductions</p></li>
<li><p>export to UCSC tracks and to loom file</p></li>
</ol>
<p>In case the process is killed prematurely, the function can be restared and the workflow will resume from the last step that was succesfully completed.</p>
<dl class="py function">
<dt id="scenicplus.wrappers.run_scenicplus.run_scenicplus">
<code class="sig-prename descclassname">scenicplus.wrappers.run_scenicplus.</code><code class="sig-name descname">run_scenicplus</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n">SCENICPLUS</span></em>, <em class="sig-param"><span class="n">variable</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">species</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">assembly</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">tf_file</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">save_path</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">biomart_host</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'http://www.ensembl.org'</span></em>, <em class="sig-param"><span class="n">upstream</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">[1000, 150000]</span></em>, <em class="sig-param"><span class="n">downstream</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">[1000, 150000]</span></em>, <em class="sig-param"><span class="n">region_ranking</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>CisTopicImputedFeatures<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gene_ranking</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>CisTopicImputedFeatures<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">calculate_TF_eGRN_correlation</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>bool<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">calculate_DEGs_DARs</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>bool<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">export_to_loom_file</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>bool<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">export_to_UCSC_file</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>bool<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">tree_structure</span><span class="p">:</span> <span class="n">Sequence<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">()</span></em>, <em class="sig-param"><span class="n">path_bedToBigBed</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">n_cpu</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">1</span></em>, <em class="sig-param"><span class="n">_temp_dir</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'/scratch/leuven/313/vsc31305/ray_spill'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/wrappers/run_scenicplus.html#run_scenicplus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.wrappers.run_scenicplus.run_scenicplus" title="Permalink to this definition"></a></dt>
<dd><p>Wrapper to run SCENIC+</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<cite>class::SCENICPLUS</cite>) – A SCENICPLUS object.</p></li>
<li><p><strong>variables</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – Variables to use for RSS, TF-eGRN correlation and markers.</p></li>
<li><p><strong>species</strong> (<em>str</em>) – Species from which data comes from. Possible values: ‘hsapiens’, ‘mmusculus’, ‘dmelanogaster’</p></li>
<li><p><strong>assembly</strong> (<em>str</em>) – Genome assembly to which the data was mapped. Possible values: ‘hg38’</p></li>
<li><p><strong>tf_file</strong> (<em>str</em>) – Path to file containing genes that are TFs</p></li>
<li><p><strong>save_path</strong> (<em>str</em>) – Folder in which results will be saved</p></li>
<li><p><strong>biomart_host</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to biomart host. Make sure that the host matches your genome assembly</p></li>
<li><p><strong>upstream</strong> (<em>str</em><em>, </em><em>optional</em>) – Upstream space to use for region to gene relationships</p></li>
<li><p><strong>downstream</strong> (<em>str</em><em>, </em><em>optional</em>) – Upstream space to use for region to gene relationships</p></li>
<li><p><strong>region_ranking</strong> (<cite>class::CisTopicImputedFeatures</cite>, optional) – Precomputed region ranking</p></li>
<li><p><strong>gene_ranking</strong> (<cite>class::CisTopicImputedFeatures</cite>, optional) – Precomputed gene ranking</p></li>
<li><p><strong>calculate_TF_eGRN_correlation</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to calculate the TF-eGRN correlation based on the variables</p></li>
<li><p><strong>calculate_DEGs_DARs</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to calculate DARs/DEGs based on the variables</p></li>
<li><p><strong>export_to_loom_file</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to export data to loom files (gene based/region based)</p></li>
<li><p><strong>export_to_UCSC_file</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to export region-to-gene links and eregulons to bed files</p></li>
<li><p><strong>tree_structure</strong> (<em>sequence</em><em>, </em><em>optional</em>) – Tree structure for loom files</p></li>
<li><p><strong>path_bedToBigBed</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to convert bed files to big bed when exporting to UCSC (required if files are meant to be
used in a hub)</p></li>
<li><p><strong>n_cpu</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of cores to use</p></li>
<li><p><strong>_temp_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Temporary directory for ray</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="tools-for-non-automated-workflow">
<h2>Tools for non-automated workflow<a class="headerlink" href="#tools-for-non-automated-workflow" title="Permalink to this headline"></a></h2>
<div class="section" id="module-scenicplus.cistromes">
<span id="cistromes"></span><h3>Cistromes<a class="headerlink" href="#module-scenicplus.cistromes" title="Permalink to this headline"></a></h3>
<p>Merging, scoring and assessing TF correlations of cistromes</p>
<dl class="py function">
<dt id="scenicplus.cistromes.TF_cistrome_correlation">
<code class="sig-prename descclassname">scenicplus.cistromes.</code><code class="sig-name descname">TF_cistrome_correlation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">variable</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">use_pseudobulk</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">auc_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Cistromes_AUC'</span></em>, <em class="sig-param"><span class="n">signature_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Unfiltered'</span></em>, <em class="sig-param"><span class="n">out_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Unfiltered'</span></em>, <em class="sig-param"><span class="n">subset</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/cistromes.html#TF_cistrome_correlation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.cistromes.TF_cistrome_correlation" title="Permalink to this definition"></a></dt>
<dd><p>Get correlation between gene expression and cistrome accessibility</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code>) – A <code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code> object with pseudobulk matrices (<cite>scplus_obj.uns[‘Pseudobulk’]</cite>)</p></li>
<li><p><strong>variable</strong> (<em>str</em><em>, </em><em>optional</em>) – Variable used to create the pseudobulks. Must be a key in <cite>scplus_obj.uns[‘Pseudobulk’]</cite>.
Only required if use_pseudobulk is False.</p></li>
<li><p><strong>use_pseudobulk</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use pseudobulk matrix or actual values.</p></li>
<li><p><strong>cistromes_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key to retrieve the pseudobulk matrices.  Cistrome accessibility will be retrieved from
<cite>scplus_obj.uns[‘Pseudobulk’][variable][‘Cistromes_AUC’][cistromes_key]</cite> and
gene expression from <cite>scplus_obj.uns[‘Pseudobulk’][variable][‘Expression’]</cite>.</p></li>
<li><p><strong>out_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Ouput key. Correlations will be stored at <cite>scplus_obj.uns[‘TF_cistrome_correlation’][variable][out_key]</cite>.</p></li>
<li><p><strong>subset</strong> (<em>List</em><em>, </em><em>optional</em>) – Subset of cells to be used to calculate correlations. Default: None (All)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.cistromes.generate_pseudobulks">
<code class="sig-prename descclassname">scenicplus.cistromes.</code><code class="sig-name descname">generate_pseudobulks</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">variable</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">normalize_expression</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">auc_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Cistromes_AUC'</span></em>, <em class="sig-param"><span class="n">signature_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Unfiltered'</span></em>, <em class="sig-param"><span class="n">nr_cells</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">10</span></em>, <em class="sig-param"><span class="n">nr_pseudobulks</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">100</span></em>, <em class="sig-param"><span class="n">seed</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">555</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/cistromes.html#generate_pseudobulks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.cistromes.generate_pseudobulks" title="Permalink to this definition"></a></dt>
<dd><p>Generate pseudobulks based on the cistrome AUC matrix and gene expression</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code>) – A <code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code> object with scored cistromes (<cite>scplus_obj.uns[‘Cistromes_AUC’][cistromes_key]</cite>)</p></li>
<li><p><strong>variable</strong> (<em>str</em><em>, </em><em>optional</em>) – Variable to create pseudobulks by. It must ve a column in <cite>scplus_obj.metadata_cell</cite></p></li>
<li><p><strong>cistromes_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key to store where cistromes AUC values are stored. Cistromes AUC will retrieved from
<cite>scplus_obj.uns[‘Cistromes_AUC’][cistromes_key]</cite> and the pseudobulk matrix will be stored
at <cite>scplus_obj.uns[‘Pseudobulk’][‘Cistromes_AUC’][variable][cistromes_key]</cite> and <cite>scplus_obj.uns[‘Pseudobulk’][variable][‘Expression’]</cite></p></li>
<li><p><strong>nr_cells</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of cells to include per pseudobulk.</p></li>
<li><p><strong>nr_pseudobulks</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of pseudobulks to generate per class</p></li>
<li><p><strong>seed</strong> (<em>int</em>) – Seed to ensure that pseudobulk are reproducible.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.cistromes.merge_cistromes">
<code class="sig-prename descclassname">scenicplus.cistromes.</code><code class="sig-name descname">merge_cistromes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">cistromes_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Unfiltered'</span></em>, <em class="sig-param"><span class="n">subset</span><span class="p">:</span> <span class="n">pyranges.pyranges.PyRanges</span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/cistromes.html#merge_cistromes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.cistromes.merge_cistromes" title="Permalink to this definition"></a></dt>
<dd><p>Generate cistromes from motif enrichment tables</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code>) – A <code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code> object with motif enrichment results from pycistarget (<cite>scplus_obj.menr</cite>).
Several analyses can be included in the slot (topics/DARs/other; and different methods [Homer/DEM/cistarget]).</p></li>
<li><p><strong>cistromes_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key to store cistromes. Cistromes will stored at <cite>scplus_obj.uns[‘Cistromes’][siganture_key]</cite></p></li>
<li><p><strong>subset</strong> (<em>list</em>) – A PyRanges containing a set of regions that regions in cistromes must overlap. This is useful when
aiming for cell type specific cistromes for example (e.g. providing the cell type’s MACS peaks)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.cistromes.prune_plot">
<code class="sig-prename descclassname">scenicplus.cistromes.</code><code class="sig-name descname">prune_plot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">name</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">pseudobulk_variable</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">auc_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Cistromes_AUC'</span></em>, <em class="sig-param"><span class="n">signature_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Unfiltered'</span></em>, <em class="sig-param"><span class="n">use_pseudobulk</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">show_dot_plot</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">show_line_plot</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">color_dict</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">subset</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">seed</span><span class="o">=</span><span class="default_value">555</span></em>, <em class="sig-param"><span class="n">ax</span><span class="p">:</span> <span class="n">matplotlib.pyplot.axes</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/cistromes.html#prune_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.cistromes.prune_plot" title="Permalink to this definition"></a></dt>
<dd><p>Plot cistrome accessibility versus TF expression</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code>) – A <code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code> object.</p></li>
<li><p><strong>cistrome_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Cistrome to plot. The TF name (or followed by extended) is enough to quert.</p></li>
<li><p><strong>pseudobulk_variable</strong> (<em>str</em><em>, </em><em>optional</em>) – Key to retrieve the pseudobulk matrices.  Cistrome accessibility will be retrieved from
<cite>scplus_obj.uns[‘Pseudobulk’][pseudobulk_variable][‘Cistromes_AUC’][cistromes_key]</cite> and
gene expression from <cite>scplus_obj.uns[‘Pseudobulk’][pseudobulk_variable][‘Expression’]</cite>.</p></li>
<li><p><strong>cistromes_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key to retrieve the pseudobulk matrices.  Cistrome accessibility will be retrieved from
<cite>scplus_obj.uns[‘Pseudobulk’][pseudobulk_variable][‘Cistromes_AUC’][cistromes_key]</cite> and
gene expression from <cite>scplus_obj.uns[‘Pseudobulk’][pseudobulk_variable][‘Expression’]</cite>.</p></li>
<li><p><strong>use_pseudobulk</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use pseudobulk matrix or actual values.</p></li>
<li><p><strong>show_dot_plot</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to show dots in plot</p></li>
<li><p><strong>show_line_plot</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to show line fitting to plot</p></li>
<li><p><strong>color_dict</strong> (<em>Dict</em><em>, </em><em>optional</em>) – Color dictionary to specify colors</p></li>
<li><p><strong>subset</strong> (<em>List</em><em>, </em><em>optional</em>) – Subset of pseudobulks/cells to use</p></li>
<li><p><strong>seed</strong> (<em>int</em>) – Seed to ensure that pseudobulk are reproducible.</p></li>
<li><p><strong>ax</strong> (<em>plt.axes</em><em>, </em><em>optional</em>) – matplotlib axes to plot to.</p></li>
<li><p><strong>**kwargs</strong> – Parameters for seaborn plotting.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.cistromes.score_cistromes">
<code class="sig-prename descclassname">scenicplus.cistromes.</code><code class="sig-name descname">score_cistromes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">ranking</span><span class="p">:</span> <span class="n">pycisTopic.diff_features.CistopicImputedFeatures</span></em>, <em class="sig-param"><span class="n">cistromes_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Unfiltered'</span></em>, <em class="sig-param"><span class="n">enrichment_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'region'</span></em>, <em class="sig-param"><span class="n">auc_threshold</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.05</span></em>, <em class="sig-param"><span class="n">normalize</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">n_cpu</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">1</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/cistromes.html#score_cistromes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.cistromes.score_cistromes" title="Permalink to this definition"></a></dt>
<dd><p>Get enrichment of a region signature in cells  using AUCell (Van de Sande et al., 2020)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code>) – A <code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code> object with motif enrichment results from pycistarget (<cite>scplus_obj.menr</cite>).</p></li>
<li><p><strong>rankings</strong> (<em>CistopicImputedFeatures</em>) – A CistopicImputedFeatures object with ranking values</p></li>
<li><p><strong>cistromes_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key to store where cistromes are stored. Cistromes will retrieved from <cite>scplus_obj.uns[‘Cistromes’][siganture_key]</cite>
and AUC values will be stored in <cite>scplus_obj.uns[‘Cistromes_AUC’][cistromes_key]</cite>.</p></li>
<li><p><strong>enrichment_type</strong> (<em>str</em>) – Whether features are genes or regions</p></li>
<li><p><strong>auc_threshold</strong> (<em>float</em>) – The fraction of the ranked genome to take into account for the calculation of the Area Under the recovery Curve. Default: 0.05</p></li>
<li><p><strong>normalize</strong> (<em>bool</em>) – Normalize the AUC values to a maximum of 1.0 per regulon. Default: False</p></li>
<li><p><strong>num_workers</strong> (<em>int</em>) – The number of cores to use. Default: 1</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>Van de Sande, B., Flerin, C., Davie, K., De Waegeneer, M., Hulselmans, G., Aibar, S., … &amp; Aerts, S. (2020). A scalable SCENIC workflow for single-cell gene
regulatory network analysis. Nature Protocols, 15(7), 2247-2276.</p>
</dd></dl>

</div>
<div class="section" id="module-scenicplus.enhancer_to_gene">
<span id="enhancer-to-gene-linking"></span><h3>Enhancer-to-gene linking<a class="headerlink" href="#module-scenicplus.enhancer_to_gene" title="Permalink to this headline"></a></h3>
<p>Link enhancers to genes based on co-occurence of chromatin accessbility of the enhancer and gene expression.</p>
<p>Both linear methods (spearman or pearson correlation) and non-linear methods (random forrest or gradient boosting) are used to link enhancers to genes.</p>
<p>The correlation methods are used to seperate regions which are infered to have a positive influence on gene expression (i.e. positive correlation)
and regions which are infered to have a negative influence on gene expression (i.e. negative correlation).</p>
<dl class="py function">
<dt id="scenicplus.enhancer_to_gene.calculate_regions_to_genes_relationships">
<code class="sig-prename descclassname">scenicplus.enhancer_to_gene.</code><code class="sig-name descname">calculate_regions_to_genes_relationships</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">SCENICPLUS_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">search_space_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'search_space'</span></em>, <em class="sig-param"><span class="n">mask_expr_dropout</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">genes</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">importance_scoring_method</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'GBM'</span></em>, <em class="sig-param"><span class="n">importance_scoring_kwargs</span><span class="p">:</span> <span class="n">dict</span> <span class="o">=</span> <span class="default_value">{'learning_rate': 0.01, 'max_features': 0.1, 'n_estimators': 500}</span></em>, <em class="sig-param"><span class="n">correlation_scoring_method</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'SR'</span></em>, <em class="sig-param"><span class="n">ray_n_cpu</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">add_distance</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">key_added</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'region_to_gene'</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/enhancer_to_gene.html#calculate_regions_to_genes_relationships"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.enhancer_to_gene.calculate_regions_to_genes_relationships" title="Permalink to this definition"></a></dt>
<dd><p>Calculates region to gene relationships using non-linear regression methods and correlation</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>SCENICPLUS_obj: SCENICPLUS</strong></dt><dd><p>instance of SCENICPLUS class containing expression data and chromatin accessbility data</p>
</dd>
<dt><strong>search_space_key: str = ‘search_space’</strong></dt><dd><p>a key in SCENICPLUS_obj.uns.keys pointing to a dataframe containing the search space surounding each gene.</p>
</dd>
<dt><strong>mask_expr_dropout: bool = False</strong></dt><dd><p>Wether or not to exclude cells which have zero counts for a gene from the calculations</p>
</dd>
<dt><strong>genes: List[str] None</strong></dt><dd><p>list of genes for which to calculate region gene scores. Default is None, i.e. all genes</p>
</dd>
<dt><strong>importance_scoring_method: str = GBM</strong></dt><dd><p>method used to score region to gene importances. Available regression analysis are: ‘RF’ (Random Forrest regression), ‘ET’ (Extra Trees regression), ‘GBM’ (Gradient Boostin regression).</p>
</dd>
<dt><strong>importance_scoring_kwargs: dict = GBM_KWARGS</strong></dt><dd><p>arguments to pass to the importance scoring function</p>
</dd>
<dt><strong>correlation_scoring_method: str = SR</strong></dt><dd><p>method used to calculate region to gene correlations. Available correlation analysis are: ‘PR’ (pearson correlation), ‘SR’ (spearman correlation).</p>
</dd>
<dt><strong>ray_n_cpu: int = None</strong></dt><dd><p>number of cores to use for ray multi-processing. Does not use ray when set to None</p>
</dd>
<dt><strong>add_distance: bool = True</strong></dt><dd><p>Wether or not to return region to gene distances</p>
</dd>
<dt><strong>key_added: str = region_to_gene</strong></dt><dd><p>Key in SCENICPLUS_obj.uns under which to store region to gene links, only stores when inplace = True</p>
</dd>
<dt><strong>inplace: bool = True</strong></dt><dd><p>Wether or not store the region to gene links in the SCENICPLUS_obj, if False a pd.DataFrame will be returned.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.enhancer_to_gene.export_to_UCSC_interact">
<code class="sig-prename descclassname">scenicplus.enhancer_to_gene.</code><code class="sig-name descname">export_to_UCSC_interact</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">SCENICPLUS_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">species</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">outfile</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">region_to_gene_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">' region_to_gene'</span></em>, <em class="sig-param"><span class="n">pbm_host</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'http://www.ensembl.org'</span></em>, <em class="sig-param"><span class="n">bigbed_outfile</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">path_bedToBigBed</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">assembly</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ucsc_track_name</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'region_to_gene'</span></em>, <em class="sig-param"><span class="n">ucsc_description</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'interaction file for region to gene'</span></em>, <em class="sig-param"><span class="n">cmap_neg</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Reds'</span></em>, <em class="sig-param"><span class="n">cmap_pos</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Greens'</span></em>, <em class="sig-param"><span class="n">key_for_color</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'importance'</span></em>, <em class="sig-param"><span class="n">vmin</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">0</span></em>, <em class="sig-param"><span class="n">vmax</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">1</span></em>, <em class="sig-param"><span class="n">scale_by_gene</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">subset_for_eRegulons_regions</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">eRegulons_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'eRegulons'</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="reference internal" href="_modules/scenicplus/enhancer_to_gene.html#export_to_UCSC_interact"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.enhancer_to_gene.export_to_UCSC_interact" title="Permalink to this definition"></a></dt>
<dd><p>Exports interaction dataframe to UCSC interaction file and (optionally) UCSC bigInteract file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>SCENICPLUS_obj: SCENICPLUS</strong></dt><dd><p>An instance of class scenicplus_class.SCENICPLUS containing region to gene links in .uns.</p>
</dd>
<dt><strong>species: str</strong></dt><dd><p>Species corresponding to your datassets (e.g. hsapiens)</p>
</dd>
<dt><strong>outfile: str</strong></dt><dd><p>Path to output file</p>
</dd>
<dt><strong>region_to_gene_key: str =’ region_to_gene’</strong></dt><dd><p>Key in SCENICPLUS_obj.uns.keys() under which to find region to gene links.</p>
</dd>
<dt><strong>pbm_host:str = ‘http://www.ensembl.org’</strong></dt><dd><p>Url of biomart host relevant for your assembly.</p>
</dd>
<dt><strong>bigbed_outfile:str = None</strong></dt><dd><p>Path to which to write the bigbed output.</p>
</dd>
<dt><strong>path_bedToBigBed: str= None</strong></dt><dd><p>Path to bedToBigBed program, used to convert bed file to bigbed format.</p>
</dd>
<dt><strong>assembly: str = None</strong></dt><dd><p>String identifying the assembly of your dataset (e.g. hg39).</p>
</dd>
<dt><strong>ucsc_track_name: str = ‘region_to_gene’</strong></dt><dd><p>Name of the exported UCSC track</p>
</dd>
<dt><strong>ucsc_description: str = ‘interaction file for region to gene’</strong></dt><dd><p>Description of the exported UCSC track</p>
</dd>
<dt><strong>cmap_neg: str = ‘Reds’</strong></dt><dd><p>Matplotlib colormap used to color negative region to gene links.</p>
</dd>
<dt><strong>cmap_pos: str = ‘Greens’</strong></dt><dd><p>Matplotlib colormap used to color positive region to gene links.</p>
</dd>
<dt><strong>key_for_color: str = ‘importance’</strong></dt><dd><p>Key pointing to column in region to gene links used to map cmap colors to.</p>
</dd>
<dt><strong>vmin: int = 0</strong></dt><dd><p>vmin of region to gene link colors.</p>
</dd>
<dt><strong>vmax: int = 1</strong></dt><dd><p>vmax of region to gene link colors.</p>
</dd>
<dt><strong>scale_by_gene: bool = True</strong></dt><dd><p>Boolean specifying wether to scale importance scores of regions linking to the same gene from 0 to 1</p>
</dd>
<dt><strong>subset_for_eRegulons_regions: bool = True</strong></dt><dd><p>Boolean specifying wether or not to subset region to gene links for regions and genes in eRegulons.</p>
</dd>
<dt><strong>eRegulons_key: str = ‘eRegulons’</strong></dt><dd><p>key in SCENICPLUS_obj.uns.keys() under which to find eRegulons.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>pd.DataFrame with region to gene links formatted in the UCSC interaction format.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.enhancer_to_gene.get_search_space">
<code class="sig-prename descclassname">scenicplus.enhancer_to_gene.</code><code class="sig-name descname">get_search_space</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">SCENICPLUS_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">species</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">assembly</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">pr_annot</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">pr_chromsizes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">predefined_boundaries</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">use_gene_boundaries</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">upstream</span><span class="o">=</span><span class="default_value">[1000, 100000]</span></em>, <em class="sig-param"><span class="n">downstream</span><span class="o">=</span><span class="default_value">[1000, 100000]</span></em>, <em class="sig-param"><span class="n">extend_tss</span><span class="o">=</span><span class="default_value">[10, 10]</span></em>, <em class="sig-param"><span class="n">remove_promoters</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">biomart_host</span><span class="o">=</span><span class="default_value">'http://www.ensembl.org'</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">key_added</span><span class="o">=</span><span class="default_value">'search_space'</span></em>, <em class="sig-param"><span class="n">implode_entries</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/enhancer_to_gene.html#get_search_space"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.enhancer_to_gene.get_search_space" title="Permalink to this definition"></a></dt>
<dd><p>Get search space surrounding genes to calculate enhancer to gene links</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>SCENICPLUS_obj: SCENICPLUS</strong></dt><dd><p>a <code class="xref py py-class docutils literal notranslate"><span class="pre">pr.SCENICPLUS</span></code>.</p>
</dd>
<dt><strong>species: string, optional</strong></dt><dd><p>Name of the species (e.g. hsapiens) on whose reference genome the search space should be calculated. This will be used to retrieve gene annotations from biomart.
Annotations can also be manually provided using the parameter [pr_annot]. Default: None</p>
</dd>
<dt><strong>assembly: string, optional</strong></dt><dd><p>Name of the assembly (e.g. hg38) of the reference genome on which the search space should be calculated.
This will be used to retrieve chromosome sizes from the UCSC genome browser.
Chromosome sizes can also be manually provided using the parameter [pr_chromsizes]. Default: None</p>
</dd>
<dt><strong>pr_annot: pr.PyRanges, optional</strong></dt><dd><p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">pr.PyRanges</span></code> containing gene annotation, including Chromosome, Start, End, Strand (as ‘+’ and ‘-‘), Gene name
and Transcription Start Site. Default: None</p>
</dd>
<dt><strong>pr_chromsizes: pr.PyRanges, optional</strong></dt><dd><p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">pr.PyRanges</span></code> containing size of each chromosome, containing ‘Chromosome’, ‘Start’ and ‘End’ columns. Default: None</p>
</dd>
<dt><strong>predefined_boundaries: pr.PyRanges, optional</strong></dt><dd><p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">pr.PyRanges</span></code> containing predefined genomic domain boundaries (e.g. TAD boundaries) to use as boundaries. If
given, use_gene_boundaries will be ignored.</p>
</dd>
<dt><strong>use_gene_boundaries: bool, optional</strong></dt><dd><p>Whether to use the whole search space or stop when encountering another gene. Default: False</p>
</dd>
<dt><strong>upstream: List, optional</strong></dt><dd><p>Search space upstream. The minimum (first position) means that even if there is a gene right next to it these
bp will be taken. The second position indicates the maximum distance. Default: [1000, 100000]</p>
</dd>
<dt><strong>downstream: List, optional</strong></dt><dd><p>Search space downstream. The minimum (first position) means that even if there is a gene right next to it these
bp will be taken. The second position indicates the maximum distance. Default: [1000, 100000]</p>
</dd>
<dt><strong>extend_tss: list, optional</strong></dt><dd><p>Space around the TSS consider as promoter. Default: [10,10]</p>
</dd>
<dt><strong>remove_promoters: bool, optional</strong></dt><dd><p>Whether to remove promoters from the search space or not. Default: False</p>
</dd>
<dt><strong>implode_entries: bool, optional</strong></dt><dd><p>When a gene has multiple start/end sites it has multiple distances and gene width.
If this parameter is set to True these multiple entries per region and gene will be put in a list, generating a single entry.
If this parameter is set to False these multiple entries will be kept.</p>
</dd>
<dt><strong>Return</strong></dt><dd></dd>
<dt><strong>——</strong></dt><dd></dd>
<dt><strong>pd.DataFrame</strong></dt><dd><p>A data frame containing regions in the search space for each gene</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-scenicplus.TF_to_gene">
<span id="tf-to-gene-linking"></span><h3>TF-to-gene linking<a class="headerlink" href="#module-scenicplus.TF_to_gene" title="Permalink to this headline"></a></h3>
<p>Link transcription factors (TFs) to genes based on co-expression of TF and target genes.</p>
<p>Both linear methods (spearman or pearson correlation) and non-linear methods (random forrest or gradient boosting) are used to link TF to genes.</p>
<p>The correlation methods are used to seperate TFs which are infered to have a positive influence on gene expression (i.e. positive correlation)
and TFs which are infered to have a negative influence on gene expression (i.e. negative correlation).</p>
<dl class="py function">
<dt id="scenicplus.TF_to_gene.calculate_TFs_to_genes_relationships">
<code class="sig-prename descclassname">scenicplus.TF_to_gene.</code><code class="sig-name descname">calculate_TFs_to_genes_relationships</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">tf_file</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">method</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'GBM'</span></em>, <em class="sig-param"><span class="n">ray_n_cpu</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">1</span></em>, <em class="sig-param"><span class="n">key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'TF2G_adj'</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/TF_to_gene.html#calculate_TFs_to_genes_relationships"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.TF_to_gene.calculate_TFs_to_genes_relationships" title="Permalink to this definition"></a></dt>
<dd><p>A function to calculate TF to gene relationships using arboreto and correlation</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>SCENICPLUS_obj</strong></dt><dd><p>An instance of <a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS"><code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code></a></p>
</dd>
<dt><strong>tf_file</strong></dt><dd><p>Path to a file specifying with genes are TFs</p>
</dd>
<dt><strong>method</strong></dt><dd><p>Whether to use Gradient Boosting Machines (GBM) or random forest (RF)</p>
</dd>
<dt><strong>ray_n_cpu</strong></dt><dd><p>Number of cpus to use</p>
</dd>
<dt><strong>key</strong></dt><dd><p>String specifying where in the .uns slot to store the adjacencies matrix in <a href="#id2"><span class="problematic" id="id3">:param:`SCENICPLUS_obj`</span></a>
default: “TF2G_adj”</p>
</dd>
<dt><strong>**kwargs</strong></dt><dd><p>Parameters to pass to ray.init</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.TF_to_gene.load_TF2G_adj_from_file">
<code class="sig-prename descclassname">scenicplus.TF_to_gene.</code><code class="sig-name descname">load_TF2G_adj_from_file</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">SCENICPLUS_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">f_adj</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">key</span><span class="o">=</span><span class="default_value">'TF2G_adj'</span></em>, <em class="sig-param"><span class="n">rho_threshold</span><span class="o">=</span><span class="default_value">0.03</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/TF_to_gene.html#load_TF2G_adj_from_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.TF_to_gene.load_TF2G_adj_from_file" title="Permalink to this definition"></a></dt>
<dd><p>Function to load TF2G adjacencies from file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>SCENICPLUS_obj</strong></dt><dd><p>An instance of <a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS"><code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code></a></p>
</dd>
<dt><strong>f_adj</strong></dt><dd><p>File path to TF2G adjacencies matrix</p>
</dd>
<dt><strong>inplace</strong></dt><dd><p>Boolean specifying wether or not to store adjacencies matrix in <cite>SCENICPLUS_obj</cite> under slot .uns[key].
Default: True</p>
</dd>
<dt><strong>key_added</strong></dt><dd><p>String specifying where in the .uns slot to store the adjacencies matrix in <cite>SCENICPLUS_obj</cite>
Default: “TF2G_adj”</p>
</dd>
<dt><strong>rho_threshold</strong></dt><dd><p>A floating point number specifying from which absolute value to consider a correlation coefficient positive or negative.
Default: 0.03</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="egrn-building">
<h3>eGRN building<a class="headerlink" href="#egrn-building" title="Permalink to this headline"></a></h3>
<div class="section" id="module-scenicplus.grn_builder.modules">
<span id="eregulon-class"></span><h4>eRegulon Class<a class="headerlink" href="#module-scenicplus.grn_builder.modules" title="Permalink to this headline"></a></h4>
<p>eRegulon class stores the transcription factor together with its target regions and genes.</p>
<dl class="py function">
<dt id="scenicplus.grn_builder.modules.create_emodules">
<code class="sig-prename descclassname">scenicplus.grn_builder.modules.</code><code class="sig-name descname">create_emodules</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">SCENICPLUS_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">region_to_gene_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'region_to_gene'</span></em>, <em class="sig-param"><span class="n">cistromes_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Unfiltered'</span></em>, <em class="sig-param"><span class="n">order_regions_to_genes_by</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'importance'</span></em>, <em class="sig-param"><span class="n">quantiles</span><span class="p">:</span> <span class="n">tuple</span> <span class="o">=</span> <span class="default_value">0.85, 0.9</span></em>, <em class="sig-param"><span class="n">top_n_regionTogenes_per_gene</span><span class="p">:</span> <span class="n">tuple</span> <span class="o">=</span> <span class="default_value">5, 10, 15</span></em>, <em class="sig-param"><span class="n">top_n_regionTogenes_per_region</span><span class="p">:</span> <span class="n">tuple</span> <span class="o">=</span> <span class="default_value">()</span></em>, <em class="sig-param"><span class="n">binarize_using_basc</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">min_regions_per_gene</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">0</span></em>, <em class="sig-param"><span class="n">rho_dichotomize</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">keep_only_activating</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">rho_threshold</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.03</span></em>, <em class="sig-param"><span class="n">keep_extended_motif_annot</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">disable_tqdm</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">ray_n_cpu</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span> &#x2192; Tuple<span class="p">[</span>List<span class="p">[</span>str<span class="p">]</span><span class="p">, </span>List<span class="p">[</span><a class="reference internal" href="#scenicplus.grn_builder.modules.eRegulon" title="scenicplus.grn_builder.modules.eRegulon">scenicplus.grn_builder.modules.eRegulon</a><span class="p">]</span><span class="p">]</span><a class="reference internal" href="_modules/scenicplus/grn_builder/modules.html#create_emodules"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.grn_builder.modules.create_emodules" title="Permalink to this definition"></a></dt>
<dd><p>A function to create e_modules of :class: ~scenicplus.grn_builder.modules.eRegulon.
This function will binarize region-to-gene links using various parameters and couples these links to a TF
based on infromation in the motif enrichment (menr) field of the <code class="xref py py-attr docutils literal notranslate"><span class="pre">SCENICPLUS_obj</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>SCENICPLUS_obj</strong></dt><dd><p>An instance of :class: <cite>~scenicplus.scenicplus_class.SCENICPLUS</cite>, containing motif enrichment data under the slot .menr.</p>
</dd>
<dt><strong>region_to_gene_key</strong></dt><dd><p>A key specifying under which to find region-to-gene links in the .uns slot of <cite>SCENICPLUS_obj</cite>.
Default: “region_to_gene”</p>
</dd>
<dt><strong>cistromes_key</strong></dt><dd><p>A key specifying which cistromes to use in .uns[‘Cistromes’] slot of <cite>SCENICPLUS_obj</cite></p>
</dd>
<dt><strong>quantiles</strong></dt><dd><p>A tuple specifying the quantiles used to binarize region-to-gene links
Default: (0.85, 0.90)</p>
</dd>
<dt><strong>top_n_regionTogenes_per_gene</strong></dt><dd><p>A tuple specifying the top n region-to-gene links to take PER GENE in order to binarize region-to-gene links.
Default: (5, 10, 15)</p>
</dd>
<dt><strong>top_n_regionTogenes_per_region</strong></dt><dd><p>A tuple specifying the top n region-to-gene links to take PER REGION in order to binarize region-to-gene links.
Default: ()</p>
</dd>
<dt><strong>binarize_using_basc:</strong></dt><dd><p>A boolean specifying wether or not to binarize region-to-gene links using BASC. Hopfensitz M, et al.</p>
</dd>
<dt><strong>min_regions_per_gene:</strong></dt><dd><p>An integer specifying a lower limit on regions per gene (after binarization) to consider for further analysis.
Default: 0</p>
</dd>
<dt><strong>rho_dichotomize:</strong></dt><dd><p>A boolean specifying wether or not to split region-to-gene links based on postive/negative correlation coefficients.
default: True</p>
</dd>
<dt><strong>keep_only_activating:</strong></dt><dd><p>A boolean specifying wether or not to only retain region-to-gene links with a positive correlation coefficient.
default: False</p>
</dd>
<dt><strong>rho_threshold:</strong></dt><dd><p>A floating point number specifying from which absolute value to consider a correlation coefficient positive or negative.
default: 0.03</p>
</dd>
<dt><strong>keep_extended_motif_annot</strong></dt><dd><p>A boolean specifying wether or not keep extended motif annotations for further analysis.
default: False</p>
</dd>
<dt><strong>ray_n_cpu</strong></dt><dd><p>An integer specifying the number of parallel cores to use to binarize region to gene links
default: None (i.e. run using single core)</p>
</dd>
<dt><strong>**kwargs</strong></dt><dd><p>Extra keyword arguments passed to ray.init function</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>A set of relevant TFs, A list of :class: <cite>~scenicplus.grn_builder.modules.eRegulon.</cite></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">References</p>
<p>Hopfensitz M, et al. Multiscale binarization of gene expression data for reconstructing Boolean networks. IEEE/ACM Trans Comput Biol Bioinform. 2012;9(2):487-98.</p>
</dd></dl>

<dl class="py class">
<dt id="scenicplus.grn_builder.modules.eRegulon">
<em class="property">class </em><code class="sig-prename descclassname">scenicplus.grn_builder.modules.</code><code class="sig-name descname">eRegulon</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">transcription_factor</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">cistrome_name</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">is_extended</span><span class="p">:</span> <span class="n">bool</span></em>, <em class="sig-param"><span class="n">regions2genes</span><span class="p">:</span> <span class="n">List<span class="p">[</span>collections.namedtuple<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">context</span><span class="o">=</span><span class="default_value">frozenset({})</span></em>, <em class="sig-param"><span class="n">in_leading_edge</span><span class="p">:</span> <span class="n">List<span class="p">[</span>bool<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gsea_enrichment_score</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gsea_pval</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gsea_adj_pval</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/grn_builder/modules.html#eRegulon"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.grn_builder.modules.eRegulon" title="Permalink to this definition"></a></dt>
<dd><p>An eRegulon is a gene signature that defines the target regions and genes of a Transcription Factor (TF).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>transcription_factor</strong></dt><dd><p>A string specifying the transcription factor of the eRegulon</p>
</dd>
<dt><strong>cistrome_name</strong></dt><dd><p>A string specifying the cistrome name</p>
</dd>
<dt><strong>is_extended</strong></dt><dd><p>A boolean specifying wether the cistromes comes from an extended annotation or not</p>
</dd>
<dt><strong>regions2genes</strong></dt><dd><p>A list of named tuples containing information on region-to-gene links
(region, gene, importance score, correlation coefficient)</p>
</dd>
<dt><strong>context</strong></dt><dd><p>The context in which this eRegulon was created
(this can be different threshold, activating / repressing, …).
default: frozenset()</p>
</dd>
<dt><strong>in_leading_edge</strong></dt><dd><p>A list specifying which genes are in the leading edge of a gsea analysis.
default: None</p>
</dd>
<dt><strong>gsea_enrichment_score</strong></dt><dd><p>A float containing the enrichment score of a gsea analysis.
default: None</p>
</dd>
<dt><strong>gsea_pval</strong></dt><dd><p>A float containing the p value of a gsea analysis.
default: None</p>
</dd>
<dt><strong>gsea_adj_pval</strong></dt><dd><p>A float containing an adjusted p value of a gsea analysis.
default: None</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#scenicplus.grn_builder.modules.create_emodules" title="scenicplus.grn_builder.modules.create_emodules"><code class="xref py py-obj docutils literal notranslate"><span class="pre">scenicplus.grn_builder.modules.create_emodules</span></code></a></dt><dd></dd>
</dl>
</div>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><a class="reference internal" href="#scenicplus.grn_builder.modules.eRegulon.target_genes" title="scenicplus.grn_builder.modules.eRegulon.target_genes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">target_genes</span></code></a></dt><dd><p>Return target genes of this eRegulon.</p>
</dd>
<dt><a class="reference internal" href="#scenicplus.grn_builder.modules.eRegulon.target_regions" title="scenicplus.grn_builder.modules.eRegulon.target_regions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">target_regions</span></code></a></dt><dd><p>Return target regions of this eRegulon.</p>
</dd>
<dt><a class="reference internal" href="#scenicplus.grn_builder.modules.eRegulon.n_target_genes" title="scenicplus.grn_builder.modules.eRegulon.n_target_genes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">n_target_genes</span></code></a></dt><dd><p>Return number of target genes.</p>
</dd>
<dt><a class="reference internal" href="#scenicplus.grn_builder.modules.eRegulon.n_target_regions" title="scenicplus.grn_builder.modules.eRegulon.n_target_regions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">n_target_regions</span></code></a></dt><dd><p>Return number of target regions.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#scenicplus.grn_builder.modules.eRegulon.subset_leading_edge" title="scenicplus.grn_builder.modules.eRegulon.subset_leading_edge"><code class="xref py py-obj docutils literal notranslate"><span class="pre">subset_leading_edge</span></code></a>([inplace])</p></td>
<td><p>Subset eReglon on leading edge.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt id="scenicplus.grn_builder.modules.eRegulon.n_target_genes">
<em class="property">property </em><code class="sig-name descname">n_target_genes</code><a class="headerlink" href="#scenicplus.grn_builder.modules.eRegulon.n_target_genes" title="Permalink to this definition"></a></dt>
<dd><p>Return number of target genes.</p>
</dd></dl>

<dl class="py method">
<dt id="scenicplus.grn_builder.modules.eRegulon.n_target_regions">
<em class="property">property </em><code class="sig-name descname">n_target_regions</code><a class="headerlink" href="#scenicplus.grn_builder.modules.eRegulon.n_target_regions" title="Permalink to this definition"></a></dt>
<dd><p>Return number of target regions.</p>
</dd></dl>

<dl class="py method">
<dt id="scenicplus.grn_builder.modules.eRegulon.subset_leading_edge">
<code class="sig-name descname">subset_leading_edge</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">inplace</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/grn_builder/modules.html#eRegulon.subset_leading_edge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.grn_builder.modules.eRegulon.subset_leading_edge" title="Permalink to this definition"></a></dt>
<dd><p>Subset eReglon on leading edge.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>inplace</strong> – if set to True update eRegulon else return new eRegulon after subset.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="scenicplus.grn_builder.modules.eRegulon.target_genes">
<em class="property">property </em><code class="sig-name descname">target_genes</code><a class="headerlink" href="#scenicplus.grn_builder.modules.eRegulon.target_genes" title="Permalink to this definition"></a></dt>
<dd><p>Return target genes of this eRegulon.</p>
</dd></dl>

<dl class="py method">
<dt id="scenicplus.grn_builder.modules.eRegulon.target_regions">
<em class="property">property </em><code class="sig-name descname">target_regions</code><a class="headerlink" href="#scenicplus.grn_builder.modules.eRegulon.target_regions" title="Permalink to this definition"></a></dt>
<dd><p>Return target regions of this eRegulon.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="scenicplus.grn_builder.modules.merge_emodules">
<code class="sig-prename descclassname">scenicplus.grn_builder.modules.</code><code class="sig-name descname">merge_emodules</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">SCENICPLUS_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">e_modules</span><span class="p">:</span> <span class="n">list</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">e_modules_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'eRegulons'</span></em>, <em class="sig-param"><span class="n">rho_dichotomize</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">key_to_add</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'eRegulons'</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/grn_builder/modules.html#merge_emodules"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.grn_builder.modules.merge_emodules" title="Permalink to this definition"></a></dt>
<dd><p>Function to merge list of <a class="reference internal" href="#scenicplus.grn_builder.modules.eRegulon" title="scenicplus.grn_builder.modules.eRegulon"><code class="xref py py-class docutils literal notranslate"><span class="pre">eRegulon</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>SCENICPLUS_obj</strong></dt><dd><p>An instance of :class: <cite>~scenicplus.scenicplus_class.SCENICPLUS</cite>, containing eRegulons in slot .uns.
default: None</p>
</dd>
<dt><strong>e_modules</strong></dt><dd><p>A list of <a class="reference internal" href="#scenicplus.grn_builder.modules.eRegulon" title="scenicplus.grn_builder.modules.eRegulon"><code class="xref py py-class docutils literal notranslate"><span class="pre">eRegulon</span></code></a>.
default: None</p>
</dd>
<dt><strong>e_modules_key</strong></dt><dd><p>A string key specifying where to find a list of <a class="reference internal" href="#scenicplus.grn_builder.modules.eRegulon" title="scenicplus.grn_builder.modules.eRegulon"><code class="xref py py-class docutils literal notranslate"><span class="pre">eRegulon</span></code></a> under .uns slot of`SCENICPLUS_obj`.
default: “eRegulons”</p>
</dd>
<dt><strong>rho_dichotomize</strong></dt><dd><p>A boolean specifying wether or not to split region-to-gene links based on postive/negative correlation coefficients.
default: True</p>
</dd>
<dt><strong>key_to_add</strong></dt><dd><p>A string key specifying where to store the list of merged <a class="reference internal" href="#scenicplus.grn_builder.modules.eRegulon" title="scenicplus.grn_builder.modules.eRegulon"><code class="xref py py-class docutils literal notranslate"><span class="pre">eRegulon</span></code></a> in the .uns slot of <cite>SCENICPLUS_obj</cite>.</p>
</dd>
<dt><strong>inplace</strong></dt><dd><p>A boolean if set to True update the <cite>SCENICPLUS_obj</cite> otherwise return list of merged <a class="reference internal" href="#scenicplus.grn_builder.modules.eRegulon" title="scenicplus.grn_builder.modules.eRegulon"><code class="xref py py-class docutils literal notranslate"><span class="pre">eRegulon</span></code></a></p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="gsea-based-approach">
<h4>GSEA based approach<a class="headerlink" href="#gsea-based-approach" title="Permalink to this headline"></a></h4>
<a class="reference internal image-reference" href="_images/gsea.png"><img alt="GSEA approach" class="align-center" src="_images/gsea.png" style="width: 800px;" /></a>
<span class="target" id="module-scenicplus.grn_builder.gsea_approach"></span><p>Generate enhancer drive GRNs (eGRS) using the GSEA approach.</p>
<p>Using this approach we will test if the gene set obtained from region-to-gene links, where the regions have a high score for a motif of a certain TF
(region indicated in black in the diagram: r8, r11, r18, r20, r22, r24),
are enriched in the top of the ranking based on the TF-to-gene links of the same TF (bottom right panel).</p>
<p>Only genes, from the set, and the regions linked to these genes in the top of the ranking (i.e. leading edge) will be kept.</p>
<p>This aproach is done seperatly for positive and negative TF and region to gene links.</p>
<p>Generating following four combinations:</p>
<table class="colwidths-given docutils align-default" id="id4">
<caption><span class="caption-text">Possible eRegulons</span><a class="headerlink" href="#id4" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>TF-to-gene relationship</p></th>
<th class="head"><p>region-to-gene relationship</p></th>
<th class="head"><p>biological role</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>positive (+)</p></td>
<td><p>positive (+)</p></td>
<td><p>TF opens chromatin and activates gene expression</p></td>
</tr>
<tr class="row-odd"><td><p>positive (+)</p></td>
<td><p>negative (-)</p></td>
<td><p>When the TF is expressed the target gene is also expressed but the regions linked to the gene are closed.</p></td>
</tr>
<tr class="row-even"><td><p>negative (-)</p></td>
<td><p>positive (+)</p></td>
<td><p>When the TF is expressed the target gene is not expressed. When the target gene is expressed, regions linked to this gene are open. TF could be a chromatin closing repressor.</p></td>
</tr>
<tr class="row-odd"><td><p>negative (-)</p></td>
<td><p>negative (-)</p></td>
<td><p>When the TF is expressed the target gene is not expressed. When the target gene is expressed, regions linked to this gene are closed.</p></td>
</tr>
</tbody>
</table>
<p>Left panel indicates the TF-to-gene links (blue) and region-to-gene links (yellow).
Witdh of arrows correspond the strength of the connections based on non-linear regression.</p>
<p>Top right panel indicates regions with a high score for a motif of a TF</p>
<p>Bottom right panel shows GSEA analysis, with on the left genes ranked by TF-to-gene connection strength and on the right the gene-set obtained from region-to-gene links.
In the diagram: g2, g6, and g10 are located in the top of the TF-to-gene ranking (i.e. leading edge), only these genes and the regions linked to these genes: r20, r18, r11, r8, r24 and r22
will be kept.</p>
<dl class="py function">
<dt id="scenicplus.grn_builder.gsea_approach.build_grn">
<code class="sig-prename descclassname">scenicplus.grn_builder.gsea_approach.</code><code class="sig-name descname">build_grn</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">SCENICPLUS_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">adj_key</span><span class="o">=</span><span class="default_value">'TF2G_adj'</span></em>, <em class="sig-param"><span class="n">cistromes_key</span><span class="o">=</span><span class="default_value">'Unfiltered'</span></em>, <em class="sig-param"><span class="n">region_to_gene_key</span><span class="o">=</span><span class="default_value">'region_to_gene'</span></em>, <em class="sig-param"><span class="n">order_regions_to_genes_by</span><span class="o">=</span><span class="default_value">'importance'</span></em>, <em class="sig-param"><span class="n">order_TFs_to_genes_by</span><span class="o">=</span><span class="default_value">'importance'</span></em>, <em class="sig-param"><span class="n">gsea_n_perm</span><span class="o">=</span><span class="default_value">1000</span></em>, <em class="sig-param"><span class="n">quantiles</span><span class="o">=</span><span class="default_value">0.85, 0.9</span></em>, <em class="sig-param"><span class="n">top_n_regionTogenes_per_gene</span><span class="o">=</span><span class="default_value">5, 10, 15</span></em>, <em class="sig-param"><span class="n">top_n_regionTogenes_per_region</span><span class="o">=</span><span class="default_value">()</span></em>, <em class="sig-param"><span class="n">binarize_using_basc</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">min_regions_per_gene</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">rho_dichotomize_tf2g</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">rho_dichotomize_r2g</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">rho_dichotomize_eregulon</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">keep_only_activating</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">rho_threshold</span><span class="o">=</span><span class="default_value">0.03</span></em>, <em class="sig-param"><span class="n">NES_thr</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">adj_pval_thr</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">min_target_genes</span><span class="o">=</span><span class="default_value">5</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">key_added</span><span class="o">=</span><span class="default_value">'eRegulons'</span></em>, <em class="sig-param"><span class="n">ray_n_cpu</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">merge_eRegulons</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">keep_extended_motif_annot</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">disable_tqdm</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/grn_builder/gsea_approach.html#build_grn"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.grn_builder.gsea_approach.build_grn" title="Permalink to this definition"></a></dt>
<dd><p>Build GRN using GSEA approach</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>SCENICPLUS_obj</strong> – An instance of :class: <cite>~scenicplus.scenicplus_class.SCENICPLUS</cite></p></li>
<li><p><strong>adj_key</strong> – Key under which to find TF2G adjacencies
default: “TF2G_adj”</p></li>
<li><p><strong>region_to_gene_key</strong> – Key under which to find R2G adjacnecies
default: “region_to_gene”</p></li>
<li><p><strong>gsea_n_perm</strong> – Int specifying number of gsea permutations to run for p value calculation
default: 1000</p></li>
<li><p><strong>quantiles</strong> – A tuple specifying the quantiles used to binarize region-to-gene links
Default: (0.85, 0.90)</p></li>
<li><p><strong>top_n_regionTogenes_per_gene</strong> – A tuple specifying the top n region-to-gene links to take PER GENE in order to binarize region-to-gene links.
Default: (5, 10, 15)</p></li>
<li><p><strong>top_n_regionTogenes_per_region</strong> – A tuple specifying the top n region-to-gene links to take PER REGION in order to binarize region-to-gene links.
Default: ()</p></li>
<li><p><strong>binarize_using_basc</strong> – A boolean specifying wether or not to binarize region-to-gene links using BASC. Hopfensitz M, et al.</p></li>
<li><p><strong>min_regions_per_gene</strong> – An integer specifying a lower limit on regions per gene (after binarization) to consider for further analysis.
Default: 0</p></li>
<li><p><strong>rho_dichotomize</strong> – A boolean specifying wether or not to split region-to-gene links based on postive/negative correlation coefficients.
default: True</p></li>
<li><p><strong>keep_only_activating</strong> – A boolean specifying wether or not to only retain region-to-gene links with a positive correlation coefficient.
default: False</p></li>
<li><p><strong>rho_threshold</strong> – A floating point number specifying from which absolute value to consider a correlation coefficient positive or negative.
default: 0.03</p></li>
<li><p><strong>NES_thr</strong> – Float specifying threshold on gsea NES value
defaut: 0</p></li>
<li><p><strong>adj_pval_thr</strong> – Float specifying threshold on gsea adjusted p value
default: 1</p></li>
<li><p><strong>min_target_genes</strong> – Int specifying minumum number of target genes in leading edge
default: 5</p></li>
<li><p><strong>inplace</strong> – Boolean specifying wether to store results in <cite>SCENICPLUS_obj</cite>
default: True</p></li>
<li><p><strong>key_added</strong> – Key specifying in under which key to store result in <cite>SCENICPLUS_obj</cite>.uns
default: “eRegulons”</p></li>
<li><p><strong>ray_n_cpu</strong> – Int specifying number of cores to use
default: None</p></li>
<li><p><strong>merge_eRegulons</strong> – Boolean specifying wether to merge eRegulons form the same TF but different thresholding approaches
default: True</p></li>
<li><p><strong>keep_extended_motif_annot</strong> – A boolean specifying wether or not keep extended motif annotations for further analysis.
default: False</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments passed to <cite>ray.init</cite></p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>Hopfensitz M, et al. Multiscale binarization of gene expression data for reconstructing Boolean networks. IEEE/ACM Trans Comput Biol Bioinform. 2012;9(2):487-98.</p>
</dd></dl>

</div>
</div>
</div>
<div class="section" id="downstream-analysis-export-and-plotting">
<h2>Downstream analysis, export and plotting<a class="headerlink" href="#downstream-analysis-export-and-plotting" title="Permalink to this headline"></a></h2>
<div class="section" id="module-scenicplus.diff_features">
<span id="marker-genes-and-regions"></span><h3>Marker genes and regions<a class="headerlink" href="#module-scenicplus.diff_features" title="Permalink to this headline"></a></h3>
<p>Calculate differentially expressed genes (DEGs) and differentially accessible regions (DARs).</p>
<dl class="py function">
<dt id="scenicplus.diff_features.get_differential_features">
<code class="sig-prename descclassname">scenicplus.diff_features.</code><code class="sig-name descname">get_differential_features</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">variable</span></em>, <em class="sig-param"><span class="n">use_hvg</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>bool<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">contrast_type</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">['DARs', 'DEGs']</span></em>, <em class="sig-param"><span class="n">adjpval_thr</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">0.05</span></em>, <em class="sig-param"><span class="n">log2fc_thr</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">0.5849625007211562</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/diff_features.html#get_differential_features"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.diff_features.get_differential_features" title="Permalink to this definition"></a></dt>
<dd><p>Get DARs of DEGs given reference variable.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<cite>class::SCENICPLUS</cite>) – A SCENICPLUS object.</p></li>
<li><p><strong>variable</strong> (<em>str</em>) – Variable to compute DARs/DEGs by (has to be included in scplus_obj.metadata_cell)</p></li>
<li><p><strong>use_hvg</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use only highly variable genes/regions</p></li>
<li><p><strong>contrast_type</strong> (<em>list</em><em>, </em><em>optional</em>) – Wheter to compute DARs and/or DEGs per variable</p></li>
<li><p><strong>adjpval_thr</strong> (<em>float</em><em>, </em><em>optional</em>) – P-value threshold</p></li>
<li><p><strong>log2fc_thr</strong> – Log2FC threshold</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-scenicplus.eregulon_enrichment">
<span id="eregulon-enrichment-in-cells"></span><h3>eRegulon enrichment in cells<a class="headerlink" href="#module-scenicplus.eregulon_enrichment" title="Permalink to this headline"></a></h3>
<p>Score eRegulon target genes and regions in cells using AUC algorithm.</p>
<dl class="py function">
<dt id="scenicplus.eregulon_enrichment.binarize_AUC">
<code class="sig-prename descclassname">scenicplus.eregulon_enrichment.</code><code class="sig-name descname">binarize_AUC</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">auc_key</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'eRegulon_AUC'</span></em>, <em class="sig-param"><span class="n">out_key</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'eRegulon_AUC_thresholds'</span></em>, <em class="sig-param"><span class="n">signature_keys</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">['Gene_based', 'Region_based']</span></em>, <em class="sig-param"><span class="n">n_cpu</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">1</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/eregulon_enrichment.html#binarize_AUC"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.eregulon_enrichment.binarize_AUC" title="Permalink to this definition"></a></dt>
<dd><p>Binarize eRegulons using AUCell</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>scplus_obj: `class::SCENICPLUS`</strong></dt><dd><p>A SCENICPLUS object with eRegulons AUC.
auc_key: str, optional
Key where the AUC values are stored</p>
</dd>
<dt><strong>out_key: str, optional</strong></dt><dd><p>Key where the AUCell thresholds will be stored (in <cite>scplus_obj.uns</cite>)</p>
</dd>
<dt><strong>signature_keys: List, optional</strong></dt><dd><p>Keys to extract AUC values from. Default: [‘Gene_based’, ‘Region_based’]</p>
</dd>
<dt><strong>n_cpu: int</strong></dt><dd><p>The number of cores to use. Default: 1</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.eregulon_enrichment.get_eRegulons_as_signatures">
<code class="sig-prename descclassname">scenicplus.eregulon_enrichment.</code><code class="sig-name descname">get_eRegulons_as_signatures</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">eRegulon_metadata_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'eRegulon_metadata'</span></em>, <em class="sig-param"><span class="n">key_added</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'eRegulon_signatures'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/eregulon_enrichment.html#get_eRegulons_as_signatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.eregulon_enrichment.get_eRegulons_as_signatures" title="Permalink to this definition"></a></dt>
<dd><p>Format eRegulons for scoring</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>scplus_obj: `class::SCENICPLUS`</strong></dt><dd><p>A SCENICPLUS object with eRegulons metadata computed.</p>
</dd>
<dt><strong>eRegulons_metadata_key: str, optional</strong></dt><dd><p>Key where the eRegulon metadata is stored (in <cite>scplus_obj.uns</cite>)</p>
</dd>
<dt><strong>key_added: str, optional</strong></dt><dd><p>Key where formated signatures will be stored (in <cite>scplus_obj.uns</cite>)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.eregulon_enrichment.make_rankings">
<code class="sig-prename descclassname">scenicplus.eregulon_enrichment.</code><code class="sig-name descname">make_rankings</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">target</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'region'</span></em>, <em class="sig-param"><span class="n">seed</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">123</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/eregulon_enrichment.html#make_rankings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.eregulon_enrichment.make_rankings" title="Permalink to this definition"></a></dt>
<dd><p>A function to generate rankings per cell based on the imputed accessibility scores per region
or the gene expression per cell.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code>) – A <code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code> object with motif enrichment results from pycistarget (<cite>scplus_obj.menr</cite>).</p></li>
<li><p><strong>target</strong> (<em>str</em><em>, </em><em>optional</em>) – Whether rankings should be done based on gene expression or region accessibilty. Default: ‘region’</p></li>
<li><p><strong>seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed to ensure reproducibility of the rankings when there are ties</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.eregulon_enrichment.score_eRegulons">
<code class="sig-prename descclassname">scenicplus.eregulon_enrichment.</code><code class="sig-name descname">score_eRegulons</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">ranking</span><span class="p">:</span> <span class="n">pycisTopic.diff_features.CistopicImputedFeatures</span></em>, <em class="sig-param"><span class="n">eRegulon_signatures_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'eRegulon_signatures'</span></em>, <em class="sig-param"><span class="n">key_added</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'eRegulon_AUC'</span></em>, <em class="sig-param"><span class="n">enrichment_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'region'</span></em>, <em class="sig-param"><span class="n">auc_threshold</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.05</span></em>, <em class="sig-param"><span class="n">normalize</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">n_cpu</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">1</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/eregulon_enrichment.html#score_eRegulons"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.eregulon_enrichment.score_eRegulons" title="Permalink to this definition"></a></dt>
<dd><p>Score eRegulons using AUCell</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>scplus_obj: `class::SCENICPLUS`</strong></dt><dd><p>A SCENICPLUS object with formatted eRegulons.</p>
</dd>
<dt><strong>eRegulons_metadata_key: `class::CistopicImputedFeatures`</strong></dt><dd><p>Precomputed region/gene ranking.</p>
</dd>
<dt><strong>eRegulon_signatures_key: str, optional</strong></dt><dd><p>Key where formated signatures are stored (in <cite>scplus_obj.uns</cite>)</p>
</dd>
<dt><strong>key_added: str, optional</strong></dt><dd><p>Key where formated AUC values will be stored (in <cite>scplus_obj.uns</cite>)</p>
</dd>
<dt><strong>enrichment_type: str, optional</strong></dt><dd><p>Whether region or gene signatures are being used</p>
</dd>
<dt><strong>auc_threshold: float</strong></dt><dd><p>The fraction of the ranked genome to take into account for the calculation of the Area Under the recovery Curve. Default: 0.05</p>
</dd>
<dt><strong>normalize: bool</strong></dt><dd><p>Normalize the AUC values to a maximum of 1.0 per regulon. Default: False</p>
</dd>
<dt><strong>n_cpu: int</strong></dt><dd><p>The number of cores to use. Default: 1</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-scenicplus.dimensionality_reduction">
<span id="dimensionality-reduction"></span><h3>dimensionality reduction<a class="headerlink" href="#module-scenicplus.dimensionality_reduction" title="Permalink to this headline"></a></h3>
<p>Dimmensionality reduction and clustering based on target genes and target regions AUC.</p>
<dl class="py function">
<dt id="scenicplus.dimensionality_reduction.find_clusters">
<code class="sig-prename descclassname">scenicplus.dimensionality_reduction.</code><code class="sig-name descname">find_clusters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">auc_key</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'eRegulon_AUC'</span></em>, <em class="sig-param"><span class="n">signature_keys</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">['Gene_based', 'Region_based']</span></em>, <em class="sig-param"><span class="n">k</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">10</span></em>, <em class="sig-param"><span class="n">res</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>float<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">[0.6]</span></em>, <em class="sig-param"><span class="n">seed</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">555</span></em>, <em class="sig-param"><span class="n">scale</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>bool<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">prefix</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">''</span></em>, <em class="sig-param"><span class="n">selected_regulons</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>int<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">selected_cells</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/dimensionality_reduction.html#find_clusters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.dimensionality_reduction.find_clusters" title="Permalink to this definition"></a></dt>
<dd><p>Performing leiden cell or region clustering and add results to SCENICPLUS object’s metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<cite>class::SCENICPLUS</cite>) – A SCENICPLUS object with eRegulons AUC computed.</p></li>
<li><p><strong>auc_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key to extract AUC values from. Default: ‘eRegulon_AUC’</p></li>
<li><p><strong>signature_keys</strong> (<em>List</em><em>, </em><em>optional</em>) – Keys to extract AUC values from. Default: [‘Gene_based’, ‘Region_based’]</p></li>
<li><p><strong>k</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of neighbours in the k-neighbours graph. Default: 10</p></li>
<li><p><strong>res</strong> (<em>float</em><em>, </em><em>optional</em>) – Resolution parameter for the leiden algorithm step. Default: 0.6</p></li>
<li><p><strong>seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Seed parameter for the leiden algorithm step. Default: 555</p></li>
<li><p><strong>scale</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to scale the enrichment prior to the clustering. Default: False</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix to add to the clustering name when adding it to the correspondent metadata attribute. Default: ‘’</p></li>
<li><p><strong>selected_regulons</strong> (<em>list</em><em>, </em><em>optional</em>) – A list with selected regulons to be used for clustering. Default: None (use all regulons)</p></li>
<li><p><strong>selected_cels</strong> (<em>list</em><em>, </em><em>optional</em>) – A list with selected features cells to cluster. Default: None (use all cells)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.dimensionality_reduction.plot_eRegulon">
<code class="sig-prename descclassname">scenicplus.dimensionality_reduction.</code><code class="sig-name descname">plot_eRegulon</code><span class="sig-paren">(</span><em class="sig-param">scplus_obj: scenicplus.scenicplus_class.SCENICPLUS, reduction_name: str, auc_key: Optional[str] = 'eRegulon_AUC', signature_keys: Optional[List[str]] = ['Gene_based', 'Region_based'], normalize_tf_expression: Optional[bool] = True, cmap: Optional[Union[str, matplotlib.cm]] = &lt;matplotlib.colors.ListedColormap object&gt;, dot_size: Optional[int] = 10, alpha: Optional[Union[float, int]] = 1, scale: Optional[bool] = False, selected_regulons: Optional[List[int]] = None, selected_cells: Optional[List[str]] = None, figsize: Optional[Tuple[float, float]] = (6.4, 4.8), num_columns: Optional[int] = 3, save: Optional[str] = None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/dimensionality_reduction.html#plot_eRegulon"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.dimensionality_reduction.plot_eRegulon" title="Permalink to this definition"></a></dt>
<dd><p>Plot TF expression and eRegulon AUC (gene and region based) into dimensionality reduction.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<cite>class::SCENICPLUS</cite>) – A cisTopic object with dimensionality reductions in <cite>class::CistopicObject.projections</cite>.</p></li>
<li><p><strong>reduction_name</strong> (<em>str</em>) – Name of the dimensionality reduction to use</p></li>
<li><p><strong>auc_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key to extract AUC values from. Default: ‘eRegulon_AUC’</p></li>
<li><p><strong>signature_keys</strong> (<em>List</em><em>, </em><em>optional</em>) – Keys to extract AUC values from. Default: [‘Gene_based’, ‘Region_based’]</p></li>
<li><p><strong>normalize_tf_expression</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether logCPM normalize TF expression. Default: True</p></li>
<li><p><strong>cmap</strong> (<em>str</em><em> or </em><em>'matplotlib.cm'</em><em>, </em><em>optional</em>) – For continuous variables, color map to use for the legend color bar. Default: cm.viridis</p></li>
<li><p><strong>dot_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Dot size in the plot. Default: 10</p></li>
<li><p><strong>alpha</strong> (<em>float</em><em>, </em><em>optional</em>) – Transparency value for the dots in the plot. Default: 1</p></li>
<li><p><strong>scale</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to scale the cell-topic or topic-regions contributions prior to plotting. Default: False</p></li>
<li><p><strong>selected_regulons</strong> (<em>list</em><em>, </em><em>optional</em>) – A list with selected regulons to be used for clustering. Default: None (use all regulons)</p></li>
<li><p><strong>selected_cells</strong> (<em>list</em><em>, </em><em>optional</em>) – A list with selected features cells to cluster. Default: None (use all cells)</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the figure. If num_columns is 1, this is the size for each figure; if num_columns is above 1, this is the overall size of the figure (if keeping
default, it will be the size of each subplot in the figure). Default: (6.4, 4.8)</p></li>
<li><p><strong>num_columns</strong> (<em>int</em><em>, </em><em>optional</em>) – For multiplot figures, indicates the number of columns (the number of rows will be automatically determined based on the number of plots). Default: 1</p></li>
<li><p><strong>save</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to save plot. Default: None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.dimensionality_reduction.plot_metadata">
<code class="sig-prename descclassname">scenicplus.dimensionality_reduction.</code><code class="sig-name descname">plot_metadata</code><span class="sig-paren">(</span><em class="sig-param">scplus_obj: scenicplus.scenicplus_class.SCENICPLUS, reduction_name: str, variables: List[str], remove_nan: Optional[bool] = True, show_label: Optional[bool] = True, show_legend: Optional[bool] = False, cmap: Optional[Union[str, matplotlib.cm]] = &lt;matplotlib.colors.ListedColormap object&gt;, dot_size: Optional[int] = 10, text_size: Optional[int] = 10, alpha: Optional[Union[float, int]] = 1, seed: Optional[int] = 555, color_dictionary: Optional[Dict[str, str]] = {}, figsize: Optional[Tuple[float, float]] = (6.4, 4.8), num_columns: Optional[int] = 1, selected_cells: Optional[List[str]] = None, save: Optional[str] = None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/dimensionality_reduction.html#plot_metadata"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.dimensionality_reduction.plot_metadata" title="Permalink to this definition"></a></dt>
<dd><p>Plot categorical and continuous metadata into dimensionality reduction.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<cite>class::SCENICPLUS</cite>) – A SCENICPLUS object with dimensionality reductions.</p></li>
<li><p><strong>reduction_name</strong> (<em>str</em>) – Name of the dimensionality reduction to use</p></li>
<li><p><strong>variables</strong> (<em>list</em>) – List of variables to plot. They should be included in <cite>class::SCENICPLUS.metadata_cell</cite>.</p></li>
<li><p><strong>remove_nan</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to remove data points for which the variable value is ‘nan’. Default: True</p></li>
<li><p><strong>show_label</strong> (<em>bool</em><em>, </em><em>optional</em>) – For categorical variables, whether to show the label in the plot. Default: True</p></li>
<li><p><strong>show_legend</strong> (<em>bool</em><em>, </em><em>optional</em>) – For categorical variables, whether to show the legend next to the plot. Default: False</p></li>
<li><p><strong>cmap</strong> (<em>str</em><em> or </em><em>'matplotlib.cm'</em><em>, </em><em>optional</em>) – For continuous variables, color map to use for the legend color bar. Default: cm.viridis</p></li>
<li><p><strong>dot_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Dot size in the plot. Default: 10</p></li>
<li><p><strong>text_size</strong> (<em>int</em><em>, </em><em>optional</em>) – For categorical variables and if show_label is True, size of the labels in the plot. Default: 10</p></li>
<li><p><strong>alpha</strong> (<em>float</em><em>, </em><em>optional</em>) – Transparency value for the dots in the plot. Default: 1</p></li>
<li><p><strong>seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed used to select random colors. Default: 555</p></li>
<li><p><strong>color_dictionary</strong> (<em>dict</em><em>, </em><em>optional</em>) – A dictionary containing an entry per variable, whose values are dictionaries with variable levels as keys and corresponding colors as values.
Default: None</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the figure. If num_columns is 1, this is the size for each figure; if num_columns is above 1, this is the overall size of the figure (if keeping
default, it will be the size of each subplot in the figure). Default: (6.4, 4.8)</p></li>
<li><p><strong>num_columns</strong> (<em>int</em><em>, </em><em>optional</em>) – For multiplot figures, indicates the number of columns (the number of rows will be automatically determined based on the number of plots). Default: 1</p></li>
<li><p><strong>selected_cells</strong> (<em>list</em><em>, </em><em>optional</em>) – A list with selected cells to plot.</p></li>
<li><p><strong>save</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to save plot. Default: None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.dimensionality_reduction.run_eRegulons_pca">
<code class="sig-prename descclassname">scenicplus.dimensionality_reduction.</code><code class="sig-name descname">run_eRegulons_pca</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">scale</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>bool<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">auc_key</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'eRegulon_AUC'</span></em>, <em class="sig-param"><span class="n">signature_keys</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">['Gene_based', 'Region_based']</span></em>, <em class="sig-param"><span class="n">reduction_name</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'eRegulons_PCA'</span></em>, <em class="sig-param"><span class="n">random_state</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">555</span></em>, <em class="sig-param"><span class="n">selected_regulons</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>int<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">selected_cells</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">n_pcs</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">50</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/dimensionality_reduction.html#run_eRegulons_pca"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.dimensionality_reduction.run_eRegulons_pca" title="Permalink to this definition"></a></dt>
<dd><p>Run UMAP and add it to the dimensionality reduction dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<cite>class::SCENICPLUS</cite>) – A SCENICPLUS object with eRegulons AUC computed.</p></li>
<li><p><strong>scale</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to scale the cell-topic or topic-regions contributions prior to the dimensionality reduction. Default: False</p></li>
<li><p><strong>auc_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key to extract AUC values from. Default: ‘eRegulon_AUC’</p></li>
<li><p><strong>signature_keys</strong> (<em>List</em><em>, </em><em>optional</em>) – Keys to extract AUC values from. Default: [‘Gene_based’, ‘Region_based’]</p></li>
<li><p><strong>reduction_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Key used to store dimensionality reduction in scplud_obj.dr_cell. Default: eRegulon_AUC.</p></li>
<li><p><strong>random_state</strong> (<em>int</em><em>, </em><em>optional</em>) – Seed parameter for running UMAP. Default: 555</p></li>
<li><p><strong>selected_regulons</strong> (<em>list</em><em>, </em><em>optional</em>) – A list with selected regulons to be used for clustering. Default: None (use all regulons)</p></li>
<li><p><strong>selected_cels</strong> (<em>list</em><em>, </em><em>optional</em>) – A list with selected features cells to cluster. Default: None (use all cells)</p></li>
<li><p><strong>**kwargs</strong> – Parameters to pass to umap.UMAP.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.dimensionality_reduction.run_eRegulons_tsne">
<code class="sig-prename descclassname">scenicplus.dimensionality_reduction.</code><code class="sig-name descname">run_eRegulons_tsne</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">scale</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>bool<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">auc_key</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'eRegulon_AUC'</span></em>, <em class="sig-param"><span class="n">signature_keys</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">['Gene_based', 'Region_based']</span></em>, <em class="sig-param"><span class="n">reduction_name</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'eRegulons_tSNE'</span></em>, <em class="sig-param"><span class="n">random_state</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">555</span></em>, <em class="sig-param"><span class="n">selected_regulons</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>int<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">selected_cells</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/dimensionality_reduction.html#run_eRegulons_tsne"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.dimensionality_reduction.run_eRegulons_tsne" title="Permalink to this definition"></a></dt>
<dd><p>Run TSNE and add it to the dimensionality reduction dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<cite>class::SCENICPLUS</cite>) – A SCENICPLUS object with eRegulons AUC computed.</p></li>
<li><p><strong>scale</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to scale the enrichments prior to the dimensionality reduction. Default: False</p></li>
<li><p><strong>auc_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key to extract AUC values from. Default: ‘eRegulon_AUC’</p></li>
<li><p><strong>signature_keys</strong> (<em>List</em><em>, </em><em>optional</em>) – Keys to extract AUC values from. Default: [‘Gene_based’, ‘Region_based’]</p></li>
<li><p><strong>reduction_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Key used to store dimensionality reduction in scplud_obj.dr_cell. Default: eRegulon_AUC.</p></li>
<li><p><strong>random_state</strong> (<em>int</em><em>, </em><em>optional</em>) – Seed parameter for running UMAP. Default: 555</p></li>
<li><p><strong>selected_regulons</strong> (<em>list</em><em>, </em><em>optional</em>) – A list with selected regulons to be used for clustering. Default: None (use all regulons)</p></li>
<li><p><strong>selected_cells</strong> (<em>list</em><em>, </em><em>optional</em>) – A list with selected features cells to cluster. Default: None (use all cells)</p></li>
<li><p><strong>**kwargs</strong> – Parameters to pass to the tSNE functions.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.dimensionality_reduction.run_eRegulons_umap">
<code class="sig-prename descclassname">scenicplus.dimensionality_reduction.</code><code class="sig-name descname">run_eRegulons_umap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">scale</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>bool<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">auc_key</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'eRegulon_AUC'</span></em>, <em class="sig-param"><span class="n">signature_keys</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">['Gene_based', 'Region_based']</span></em>, <em class="sig-param"><span class="n">reduction_name</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'eRegulons_UMAP'</span></em>, <em class="sig-param"><span class="n">random_state</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">555</span></em>, <em class="sig-param"><span class="n">selected_regulons</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>int<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">selected_cells</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/dimensionality_reduction.html#run_eRegulons_umap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.dimensionality_reduction.run_eRegulons_umap" title="Permalink to this definition"></a></dt>
<dd><p>Run UMAP and add it to the dimensionality reduction dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<cite>class::SCENICPLUS</cite>) – A SCENICPLUS object with eRegulons AUC computed.</p></li>
<li><p><strong>scale</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to scale the cell-topic or topic-regions contributions prior to the dimensionality reduction. Default: False</p></li>
<li><p><strong>auc_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key to extract AUC values from. Default: ‘eRegulon_AUC’</p></li>
<li><p><strong>signature_keys</strong> (<em>List</em><em>, </em><em>optional</em>) – Keys to extract AUC values from. Default: [‘Gene_based’, ‘Region_based’]</p></li>
<li><p><strong>reduction_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Key used to store dimensionality reduction in scplud_obj.dr_cell. Default: eRegulon_AUC.</p></li>
<li><p><strong>random_state</strong> (<em>int</em><em>, </em><em>optional</em>) – Seed parameter for running UMAP. Default: 555</p></li>
<li><p><strong>selected_regulons</strong> (<em>list</em><em>, </em><em>optional</em>) – A list with selected regulons to be used for clustering. Default: None (use all regulons)</p></li>
<li><p><strong>selected_cels</strong> (<em>list</em><em>, </em><em>optional</em>) – A list with selected features cells to cluster. Default: None (use all cells)</p></li>
<li><p><strong>**kwargs</strong> – Parameters to pass to umap.UMAP.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-scenicplus.RSS">
<span id="eregulon-specificity-score-erss"></span><h3>eRegulon specificity score (eRSS)<a class="headerlink" href="#module-scenicplus.RSS" title="Permalink to this headline"></a></h3>
<p>Calculate the specificty of eRegulons in clusters of cells.</p>
<p>Calculates the distance between the real distribution of eRegulon AUC values and a fictional distribution where the eRegulon is only expressed/accessible in cells of a certain cluster.</p>
<dl class="py function">
<dt id="scenicplus.RSS.plot_rss">
<code class="sig-prename descclassname">scenicplus.RSS.</code><code class="sig-name descname">plot_rss</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">rss_key</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">top_n</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">5</span></em>, <em class="sig-param"><span class="n">selected_groups</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">num_columns</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">1</span></em>, <em class="sig-param"><span class="n">figsize</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>Tuple<span class="p">[</span>float<span class="p">, </span>float<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">6.4, 4.8</span></em>, <em class="sig-param"><span class="n">fontsize</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">12</span></em>, <em class="sig-param"><span class="n">save</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/RSS.html#plot_rss"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.RSS.plot_rss" title="Permalink to this definition"></a></dt>
<dd><p>Plot RSS values per group</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<cite>class::SCENICPLUS</cite>) – A SCENICPLUS object with eRegulons AUC computed.</p></li>
<li><p><strong>rss_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key to extract RSS values from.</p></li>
<li><p><strong>top_n</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of top eRegulons to highlight.</p></li>
<li><p><strong>selected_groups</strong> (<em>List</em><em>, </em><em>optional</em>) – Groups to plot. Default: None (all)</p></li>
<li><p><strong>num_columns</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of columns for multiplotting</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the figure. If num_columns is 1, this is the size for each figure; if num_columns is above 1, this is the overall size of the figure (if keeping
default, it will be the size of each subplot in the figure). Default: (6.4, 4.8)</p></li>
<li><p><strong>fontsize</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of the eRegulons names in plot.</p></li>
<li><p><strong>save</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to save plot. Default: None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.RSS.regulon_specificity_scores">
<code class="sig-prename descclassname">scenicplus.RSS.</code><code class="sig-name descname">regulon_specificity_scores</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">variable</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">auc_key</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'eRegulon_AUC'</span></em>, <em class="sig-param"><span class="n">signature_keys</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">['Gene_based', 'Region_based']</span></em>, <em class="sig-param"><span class="n">selected_regulons</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>int<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">scale</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>bool<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">out_key_suffix</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">''</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/RSS.html#regulon_specificity_scores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.RSS.regulon_specificity_scores" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the Regulon Specificty Scores (RSS). [doi: 10.1016/j.celrep.2018.10.045]</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<cite>class::SCENICPLUS</cite>) – A SCENICPLUS object with eRegulons AUC computed.</p></li>
<li><p><strong>variable</strong> (<em>str</em>) – Variable to calculate the RSS values for.</p></li>
<li><p><strong>auc_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key to extract AUC values from. Default: ‘eRegulon_AUC’</p></li>
<li><p><strong>signature_keys</strong> (<em>List</em><em>, </em><em>optional</em>) – Keys to extract AUC values from. Default: [‘Gene_based’, ‘Region_based’]</p></li>
<li><p><strong>scale</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to scale the enrichment prior to the clustering. Default: False</p></li>
<li><p><strong>out_key_suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – Suffix to add to the variable name to store the values (at scplus_obj.uns[‘RSS’])</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-scenicplus.networks">
<span id="network"></span><h3>Network<a class="headerlink" href="#module-scenicplus.networks" title="Permalink to this headline"></a></h3>
<p>export eRegulons to eGRN network and plot.</p>
<dl class="py function">
<dt id="scenicplus.networks.concentrical_layout">
<code class="sig-prename descclassname">scenicplus.networks.</code><code class="sig-name descname">concentrical_layout</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">G</span></em>, <em class="sig-param"><span class="n">dist_genes</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">dist_TF</span><span class="o">=</span><span class="default_value">0.1</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/networks.html#concentrical_layout"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.networks.concentrical_layout" title="Permalink to this definition"></a></dt>
<dd><p>Generate custom concentrical layout</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>G</strong> (<em>Graph</em>) – A networkx graph</p></li>
<li><p><strong>dist_genes</strong> (<em>int</em><em>, </em><em>optional</em>) – Distance from the regions to the genes</p></li>
<li><p><strong>dist_TF</strong> – Distance from the TF to the regions</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.networks.create_nx_graph">
<code class="sig-prename descclassname">scenicplus.networks.</code><code class="sig-name descname">create_nx_graph</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">nx_tables</span><span class="p">:</span> <span class="n">Dict</span></em>, <em class="sig-param"><span class="n">use_edge_tables</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="default_value">['TF2R', 'R2G']</span></em>, <em class="sig-param"><span class="n">color_edge_by</span><span class="p">:</span> <span class="n">Dict</span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">transparency_edge_by</span><span class="p">:</span> <span class="n">Dict</span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">width_edge_by</span><span class="p">:</span> <span class="n">Dict</span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">color_node_by</span><span class="p">:</span> <span class="n">Dict</span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">transparency_node_by</span><span class="p">:</span> <span class="n">Dict</span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">size_node_by</span><span class="p">:</span> <span class="n">Dict</span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">shape_node_by</span><span class="p">:</span> <span class="n">Dict</span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">label_size_by</span><span class="p">:</span> <span class="n">Dict</span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">label_color_by</span><span class="p">:</span> <span class="n">Dict</span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">layout</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'concentrical_layout'</span></em>, <em class="sig-param"><span class="n">lc_dist_genes</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.8</span></em>, <em class="sig-param"><span class="n">lc_dist_TF</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">scale_position_by</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">250</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/networks.html#create_nx_graph"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.networks.create_nx_graph" title="Permalink to this definition"></a></dt>
<dd><p>Format node/edge feature tables into a graph</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nx_tables</strong> (<em>Dict</em>) – Dictionary with node/edge feature tables as produced by <cite>create_nx_tables</cite></p></li>
<li><p><strong>use_edge_tables</strong> (<em>List</em><em>, </em><em>optional</em>) – List of edge tables to use</p></li>
<li><p><strong>color_edge_by</strong> (<em>Dict</em><em>, </em><em>optional</em>) – A dictionary containing for a given edge key the variable and color map to color edges by.
If the variable is categorical, the entry ‘categorical_color’ can be provided as a dictionary with
category: color. If it is a continuous variable a color map can be provided as <cite>continuous_color</cite> and
entried v_max and v_min can be provided to control the min and max values of the scale. Alternatively,
one fixed color can use by using ‘fixed_color’ as variable, alterntively adding an entry fixed_color: color
to the dictionary.</p></li>
<li><p><strong>transparency_edge_by</strong> (<em>Dict</em><em>, </em><em>optional</em>) – A dictionary containing for a given edge key the variable and the max and min alpha values. The variable
name has to be provided (only continuous variables accepted), together with v_max/v_mix parameters if
desired. Alternatively, one fixed alpha can use by using ‘fixed_alpha’ as variable, alterntively adding an
entry fixed_alpha: size to the dictionary.</p></li>
<li><p><strong>width_edge_by</strong> (<em>Dict</em><em>, </em><em>optional</em>) – A dictionary containing for a given edge key the variable and the max and min sizes. The variable
name has to be provided (only continuous variables accepted), together with max_size/min_size parameters if
desired. Alternatively, one fixed size can use by using ‘fixed_size’ as variable, alterntively adding an
entry fixed_size: size to the dictionary.</p></li>
<li><p><strong>color_node_by</strong> (<em>Dict</em><em>, </em><em>optional</em>) – A dictionary containing for a given node key the variable and color map to color edges by.
If the variable is categorical, the entry ‘categorical_color’ can be provided as a dictionary with
category: color. If it is a continuous variable a color map can be provided as <cite>continuous_color</cite> and
entried v_max and v_min can be provided to control the min and max values of the scale. Alternatively,
one fixed color can use by using ‘fixed_color’ as variable, alterntively adding an entry fixed_color: color
to the dictionary.</p></li>
<li><p><strong>transparency_node_by</strong> (<em>Dict</em><em>, </em><em>optional</em>) – A dictionary containing for a given node key the variable and the max and min alpha values. The variable
name has to be provided (only continuous variables accepted), together with v_max/v_mix parameters if
desired. Alternatively, one fixed alpha can use by using ‘fixed_alpha’ as variable, alterntively adding an
entry fixed_alpha: size to the dictionary.</p></li>
<li><p><strong>size_node_by</strong> (<em>Dict</em><em>, </em><em>optional</em>) – A dictionary containing for a given node key the variable and the max and min sizes. The variable
name has to be provided (only continuous variables accepted), together with max_size/min_size parameters if
desired. Alternatively, one fixed size can use by using ‘fixed_size’ as variable, alterntively adding an
entry fixed_size: size to the dictionary.</p></li>
<li><p><strong>shape_node_by</strong> (<em>Dict</em><em>, </em><em>optional</em>) – A dictionary containing for a given node key the variable and shapes. The variable
name has to be provided (only categorical variables accepted). Alternatively, one fixed shape can use by
using ‘fixed_shape’ as variable, alterntively adding an entry fixed_shape: size to the dictionary.</p></li>
<li><p><strong>label_size_node_by</strong> (<em>Dict</em><em>, </em><em>optional</em>) – A dictionary containing for a given node key the variable and the max and min sizes. The variable
name has to be provided (only continuous variables accepted), together with max_size/min_size parameters if
desired. Alternatively, one fixed size can use by using ‘fixed_label_size’ as variable, alterntively adding an
entry fixed_label_size: size to the dictionary.</p></li>
<li><p><strong>label_color_node_by</strong> (<em>Dict</em><em>, </em><em>optional</em>) – A dictionary containing for a given node key the variable and a color dictionary. The variable
name has to be provided (only categorical variables accepted), together with a color dictionary if
desired. Alternatively, one fixed color can use by using ‘fixed_label_color’ as variable, alterntively adding an
entry fixed_label_color: size to the dictionary.</p></li>
<li><p><strong>layout</strong> (<em>str</em><em>, </em><em>optional</em>) – Layout to use. Options are: ‘concentrical_layout’ (SCENIC+ custom layout) or kamada_kawai_layout (from networkx).</p></li>
<li><p><strong>lc_dist_genes</strong> (<em>float</em><em>, </em><em>optional</em>) – Distance between regions and genes. Only used if using concentrical_layout.</p></li>
<li><p><strong>lc_dist_TF</strong> (<em>float</em><em>, </em><em>optional</em>) – Distance between TF and regions. Only used if using concentrical_layout.</p></li>
<li><p><strong>scale_position_by</strong> (<em>int</em><em>, </em><em>optional</em>) – Value to scale positions for visualization in pyvis.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.networks.create_nx_tables">
<code class="sig-prename descclassname">scenicplus.networks.</code><code class="sig-name descname">create_nx_tables</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n">SCENICPLUS</span></em>, <em class="sig-param"><span class="n">eRegulon_metadata_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'eRegulon_metadata'</span></em>, <em class="sig-param"><span class="n">subset_eRegulons</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">subset_regions</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">subset_genes</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">add_differential_gene_expression</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">add_differential_region_accessibility</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">differential_variable</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/networks.html#create_nx_tables"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.networks.create_nx_tables" title="Permalink to this definition"></a></dt>
<dd><p>A function to format eRegulon data into tables for plotting eGRNs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS"><em>SCENICPLUS</em></a>) – A SCENICPLUS object with eRegulons</p></li>
<li><p><strong>eRegulon_metadata_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key where the eRegulon metadata dataframe is stored</p></li>
<li><p><strong>subset_eRegulons</strong> (<em>list</em><em>, </em><em>optional</em>) – List of eRegulons to subset</p></li>
<li><p><strong>subset_regions</strong> (<em>list</em><em>, </em><em>optional</em>) – List of regions to subset</p></li>
<li><p><strong>subset_genes</strong> (<em>list</em><em>, </em><em>optional</em>) – List of genes to subset</p></li>
<li><p><strong>add_differential_gene_expression</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to calculate differential gene expression logFC for a given variable</p></li>
<li><p><strong>add_differential_region_accessibility</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to calculate differential region accessibility logFC for a given variable</p></li>
<li><p><strong>differential_variable</strong> (<em>list</em><em>, </em><em>optional</em>) – Variable to calculate differential gene expression or region accessibility.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.networks.plot_networkx">
<code class="sig-prename descclassname">scenicplus.networks.</code><code class="sig-name descname">plot_networkx</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">G</span></em>, <em class="sig-param"><span class="n">pos</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/networks.html#plot_networkx"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.networks.plot_networkx" title="Permalink to this definition"></a></dt>
<dd><p>A function to plot networks with networkx</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>G</strong> (<em>Graph</em>) – A networkx graph</p></li>
<li><p><strong>pos</strong> (<em>Dict</em>) – Position values</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-scenicplus.plotting.correlation_plot">
<span id="correlation-plot"></span><h3>Correlation plot<a class="headerlink" href="#module-scenicplus.plotting.correlation_plot" title="Permalink to this headline"></a></h3>
<p>Plot correlation and overlap of eRegulons</p>
<dl class="py function">
<dt id="scenicplus.plotting.correlation_plot.correlation_heatmap">
<code class="sig-prename descclassname">scenicplus.plotting.correlation_plot.</code><code class="sig-name descname">correlation_heatmap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">auc_key</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'eRegulon_AUC'</span></em>, <em class="sig-param"><span class="n">signature_keys</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">['Gene_based', 'Region_based']</span></em>, <em class="sig-param"><span class="n">scale</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>bool<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">linkage_method</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'average'</span></em>, <em class="sig-param"><span class="n">fcluster_threshold</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">selected_regulons</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>int<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">cmap</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'viridis'</span></em>, <em class="sig-param"><span class="n">plotly_height</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">1000</span></em>, <em class="sig-param"><span class="n">fontsize</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">3</span></em>, <em class="sig-param"><span class="n">save</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">use_plotly</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">figsize</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>Tuple<span class="p">[</span>int<span class="p">, </span>int<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">20, 20</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/plotting/correlation_plot.html#correlation_heatmap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.plotting.correlation_plot.correlation_heatmap" title="Permalink to this definition"></a></dt>
<dd><p>Plot correlation between eRegulons enrichment,</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<cite>class::SCENICPLUS</cite>) – A SCENICPLUS object with eRegulons AUC computed.</p></li>
<li><p><strong>auc_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key to extract AUC values from. Default: ‘eRegulon_AUC’</p></li>
<li><p><strong>signature_keys</strong> (<em>List</em><em>, </em><em>optional</em>) – Keys to extract AUC values from. Default: [‘Gene_based’, ‘Region_based’]</p></li>
<li><p><strong>scale</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to scale the enrichments prior to the dimensionality reduction. Default: False</p></li>
<li><p><strong>linkage_method</strong> (<em>str</em><em>, </em><em>optional</em>) – Linkage method to use for clustering. See <cite>scipy.cluster.hierarchy.linkage</cite>.</p></li>
<li><p><strong>fcluster_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Threshold to use to divide hierarchical clustering into clusters. See <cite>scipy.cluster.hierarchy.fcluster</cite>.</p></li>
<li><p><strong>selected_regulons</strong> (<em>list</em><em>, </em><em>optional</em>) – A list with selected regulons to be used for clustering. Default: None (use all regulons)</p></li>
<li><p><strong>cmap</strong> (<em>str</em><em> or </em><em>'matplotlib.cm'</em><em>, </em><em>optional</em>) – For continuous variables, color map to use for the legend color bar. Default: cm.viridis</p></li>
<li><p><strong>plotly_height</strong> (<em>int</em><em>, </em><em>optional</em>) – Height of the plotly plot. Width will be adjusted accordingly</p></li>
<li><p><strong>fontsize</strong> (<em>int</em><em>, </em><em>optional</em>) – Labels fontsize</p></li>
<li><p><strong>save</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to save heatmap as file</p></li>
<li><p><strong>use_plotly</strong> (<em>bool</em><em>, </em><em>optional</em>) – Use plotly or seaborn to generate the image</p></li>
<li><p><strong>figsize</strong> (<em>tupe</em><em>, </em><em>optional</em>) – Matplotlib figsize, used only if use_plotly == False</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.plotting.correlation_plot.jaccard_heatmap">
<code class="sig-prename descclassname">scenicplus.plotting.correlation_plot.</code><code class="sig-name descname">jaccard_heatmap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">gene_or_region_based</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Gene_based'</span></em>, <em class="sig-param"><span class="n">signature_key</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'eRegulon_signatures'</span></em>, <em class="sig-param"><span class="n">selected_regulons</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>int<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">linkage_method</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'average'</span></em>, <em class="sig-param"><span class="n">fcluster_threshold</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>float<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">cmap</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'viridis'</span></em>, <em class="sig-param"><span class="n">plotly_height</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">1000</span></em>, <em class="sig-param"><span class="n">fontsize</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">3</span></em>, <em class="sig-param"><span class="n">save</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">use_plotly</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">figsize</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>Tuple<span class="p">[</span>int<span class="p">, </span>int<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">20, 20</span></em>, <em class="sig-param"><span class="n">vmin</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">vmax</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/plotting/correlation_plot.html#jaccard_heatmap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.plotting.correlation_plot.jaccard_heatmap" title="Permalink to this definition"></a></dt>
<dd><p>Plot jaccard index of regions/genes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<cite>class::SCENICPLUS</cite>) – A SCENICPLUS object with eRegulon signatures.</p></li>
<li><p><strong>gene_or_region_based</strong> (<em>str</em>) – Gene_based or Region_based eRegulon signatures to use.</p></li>
<li><p><strong>signature_key</strong> (<em>List</em><em>, </em><em>optional</em>) – Key to extract eRegulon signatures from</p></li>
<li><p><strong>selected_regulons</strong> (<em>list</em><em>, </em><em>optional</em>) – A list with selected regulons to be used for clustering. Default: None (use all regulons)</p></li>
<li><p><strong>linkage_method</strong> (<em>str</em><em>, </em><em>optional</em>) – Linkage method to use for clustering. See <cite>scipy.cluster.hierarchy.linkage</cite>.</p></li>
<li><p><strong>fcluster_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Threshold to use to divide hierarchical clustering into clusters. See <cite>scipy.cluster.hierarchy.fcluster</cite>.</p></li>
<li><p><strong>cmap</strong> (<em>str</em><em> or </em><em>'matplotlib.cm'</em><em>, </em><em>optional</em>) – For continuous variables, color map to use for the legend color bar. Default: cm.viridis</p></li>
<li><p><strong>plotly_height</strong> (<em>int</em><em>, </em><em>optional</em>) – Height of the plotly plot. Width will be adjusted accordingly</p></li>
<li><p><strong>fontsize</strong> (<em>int</em><em>, </em><em>optional</em>) – Labels fontsize</p></li>
<li><p><strong>save</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to save heatmap as file</p></li>
<li><p><strong>use_plotly</strong> (<em>bool</em><em>, </em><em>optional</em>) – Use plotly or seaborn to generate the image</p></li>
<li><p><strong>figsize</strong> (<em>tupe</em><em>, </em><em>optional</em>) – Matplotlib figsize, used only if use_plotly == False</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-scenicplus.plotting.coverageplot">
<span id="coverage-plot"></span><h3>Coverage plot<a class="headerlink" href="#module-scenicplus.plotting.coverageplot" title="Permalink to this headline"></a></h3>
<p>Plot chromatin accessibility profiles and region to gene arcs.</p>
<dl class="py function">
<dt id="scenicplus.plotting.coverageplot.coverage_plot">
<code class="sig-prename descclassname">scenicplus.plotting.coverageplot.</code><code class="sig-name descname">coverage_plot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">SCENICPLUS_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">bw_dict</span><span class="p">:</span> <span class="n">Mapping<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">region</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">genes_violin_plot</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">genes_arcs</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gene_height</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">1</span></em>, <em class="sig-param"><span class="n">exon_height</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">4</span></em>, <em class="sig-param"><span class="n">meta_data_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">pr_consensus_bed</span><span class="p">:</span> <span class="n">pyranges.pyranges.PyRanges</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">region_bed_height</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">1</span></em>, <em class="sig-param"><span class="n">pr_gtf</span><span class="p">:</span> <span class="n">pyranges.pyranges.PyRanges</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">pr_interact</span><span class="p">:</span> <span class="n">pyranges.pyranges.PyRanges</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">bw_ymax</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">color_dict</span><span class="p">:</span> <span class="n">Mapping<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">cmap</span><span class="o">=</span><span class="default_value">'tab20'</span></em>, <em class="sig-param"><span class="n">plot_order</span><span class="p">:</span> <span class="n">list</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">figsize</span><span class="p">:</span> <span class="n">tuple</span> <span class="o">=</span> <span class="default_value">6, 8</span></em>, <em class="sig-param"><span class="n">fontsize_dict</span><span class="o">=</span><span class="default_value">{'bigwig_label': 9, 'bigwig_tick_label': 5, 'gene_label': 9, 'title': 12, 'violinplots_xlabel': 9, 'violinplots_ylabel': 9}</span></em>, <em class="sig-param"><span class="n">gene_label_offset</span><span class="o">=</span><span class="default_value">3</span></em>, <em class="sig-param"><span class="n">arc_rad</span><span class="o">=</span><span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">arc_lw</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">cmap_violinplots</span><span class="o">=</span><span class="default_value">'Greys'</span></em>, <em class="sig-param"><span class="n">violinplots_means_color</span><span class="o">=</span><span class="default_value">'black'</span></em>, <em class="sig-param"><span class="n">violinplots_edge_color</span><span class="o">=</span><span class="default_value">'black'</span></em>, <em class="sig-param"><span class="n">violoinplots_alpha</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">width_ratios_dict</span><span class="o">=</span><span class="default_value">{'bigwig': 3, 'violinplots': 1}</span></em>, <em class="sig-param"><span class="n">height_ratios_dict</span><span class="o">=</span><span class="default_value">{'arcs': 5, 'bigwig_violin': 1, 'custom_ax': 2, 'genes': 0.5}</span></em>, <em class="sig-param"><span class="n">sort_vln_plots</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">add_custom_ax</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="reference internal" href="_modules/scenicplus/plotting/coverageplot.html#coverage_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.plotting.coverageplot.coverage_plot" title="Permalink to this definition"></a></dt>
<dd><p>Inspired by: <a class="reference external" href="https://satijalab.org/signac/reference/coverageplot">https://satijalab.org/signac/reference/coverageplot</a></p>
<p>Generates a figure showing chromatin accessibility coverage tracks, gene expression violin plots and region to gene interactions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>SCENICPLUS_obj</strong></dt><dd><p>An instance of class <cite>~scenicplus.scenicplus_class.SCENICPLUS</cite>.</p>
</dd>
<dt><strong>bw_dict</strong></dt><dd><p>A dict containing celltype annotations/cell groups as keys and paths to bigwig files as values.</p>
</dd>
<dt><strong>region</strong></dt><dd><p>A string specifying the region of interest, in chr:start-end format.</p>
</dd>
<dt><strong>genes_violin_plot</strong></dt><dd><p>A list or string specifying for which gene(s) to plot gene expression violin plots.
default: None</p>
</dd>
<dt><strong>genes_arcs</strong></dt><dd><p>A list or string specifying for which gene(s) to plot arcs
default: None (i.e. all genes in window)</p>
</dd>
<dt><strong>gene_height</strong></dt><dd><p>An int specifying the size of non-exon parts of a gene (shown underneath the coverage plot).
default: 1</p>
</dd>
<dt><strong>exon_height</strong></dt><dd><p>An int specifying the size of nexon parts of a gene (shown underneath the coverage plot).
default: 2</p>
</dd>
<dt><strong>meta_data_key</strong></dt><dd><p>A key specifying were to find annotations corresponding to the keys in <cite>bw_dict</cite> in <cite>SCENICPLUS_obj.metadata_cell</cite>
default: None</p>
</dd>
<dt><strong>pr_consensus_bed</strong></dt><dd><p>An instance of class <cite>pyranges.PyRanges</cite> containing consensus peaks. Names of these peaks should correspond to annotations.
default: None</p>
</dd>
<dt><strong>region_bed_height</strong></dt><dd><p>An int specifying with which height to draw lines corresponding to regions in <cite>pr_consensus_bed</cite>
default: 1</p>
</dd>
<dt><strong>pr_gtf</strong></dt><dd><p>An instance of class <cite>pyranges.PyRanges</cite> containing gtf-formatted information about gene structure.
default: None</p>
</dd>
<dt><strong>pr_interact</strong></dt><dd><p>An instance of class <cite>pyranges.PyRanges</cite> containing region to gene link information in ucsc interact format (use <cite>scenicplus.utils.get_interaction_pr</cite> to create such an object).
default: None</p>
</dd>
<dt><strong>bw_ymax</strong></dt><dd><p>A float specifying the maximum height at which to draw coverage plots.
default: None</p>
</dd>
<dt><strong>color_dict</strong></dt><dd><p>A dict specifying colors for each key in <cite>bw_dict</cite>.
default: None</p>
</dd>
<dt><strong>cmap</strong></dt><dd><p>A string specifying a <cite>matplotlib.cm</cite> colormap to use for coloring coverage plots.
default: ‘tab20’</p>
</dd>
<dt><strong>plot_order</strong></dt><dd><p>A list specifying the order to use for plotting coverage plots and violin plots.
default: None</p>
</dd>
<dt><strong>figsize</strong></dt><dd><p>A tuple specifying the fig size/
default: (6, 8)</p>
</dd>
<dt><strong>fontsize_dict</strong></dt><dd><p>A dictionary specifying the fontsize of various labels in the plot.
default: {‘bigwig_label’: 9, ‘gene_label’: 9, ‘violinplots_xlabel’: 9, ‘title’: 12, ‘bigwig_tick_label’: 5}</p>
</dd>
<dt><strong>gene_label_offset</strong></dt><dd><p>The y-offset of the label underneath each gene.
default: 3</p>
</dd>
<dt><strong>arc_rad</strong></dt><dd><p>The amount of radians the region-to-gene arc should be bend.
default: 0.5</p>
</dd>
<dt><strong>arc_lw</strong></dt><dd><p>The linewidth of the region-to-gene arcs.
default: 1</p>
</dd>
<dt><strong>cmap_violinplots</strong></dt><dd><p>A string specifying a <cite>matplotlib.cm</cite> colormap to use for coloring violinplots.
default: ‘Greys’</p>
</dd>
<dt><strong>violinplots_means_color</strong></dt><dd><p>A string specifying which color to use to indicate the mean value in violinplots.
default: ‘black’</p>
</dd>
<dt><strong>violinplots_edge_color</strong></dt><dd><p>A string specifying which color to use for the edge of the violinplots.
default: ‘black’</p>
</dd>
<dt><strong>violoinplots_alpha</strong></dt><dd><p>The alpha value of the violinplots facecolor.
default: 1</p>
</dd>
<dt><strong>width_ratios_dict</strong></dt><dd><p>A dict specifying the ratio in vertical direction each part of the plot should use.
default: {‘bigwig’: 3, ‘violinplots’: 1}</p>
</dd>
<dt><strong>height_ratios_dict</strong></dt><dd><p>A dict specifying the ratio in horizontal direction each part of the plot should use.
default: {‘bigwig_violin’: 1, ‘genes’: 0.5, ‘arcs’: 5}</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-scenicplus.plotting.dotplot">
<span id="dotplot"></span><h3>dotplot<a class="headerlink" href="#module-scenicplus.plotting.dotplot" title="Permalink to this headline"></a></h3>
<p>Plot TF expression, motif enrichment and AUC values of target genes and regions in a dotplot.</p>
<dl class="py function">
<dt id="scenicplus.plotting.dotplot.dotplot">
<code class="sig-prename descclassname">scenicplus.plotting.dotplot.</code><code class="sig-name descname">dotplot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">df_dotplot</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">ax</span><span class="p">:</span> <span class="n">matplotlib.pyplot.axes</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">region_set_key</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Name'</span></em>, <em class="sig-param"><span class="n">size_var</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'TF_expression'</span></em>, <em class="sig-param"><span class="n">color_var</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Cistrome_AUC'</span></em>, <em class="sig-param"><span class="n">order_group</span><span class="p">:</span> <span class="n">list</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">order_cistromes</span><span class="p">:</span> <span class="n">list</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">order_cistromes_by_max</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'TF_expression'</span></em>, <em class="sig-param"><span class="n">cluster</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">n_clust</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">2</span></em>, <em class="sig-param"><span class="n">min_point_size</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">3</span></em>, <em class="sig-param"><span class="n">max_point_size</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">30</span></em>, <em class="sig-param"><span class="n">cmap</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'viridis'</span></em>, <em class="sig-param"><span class="n">vmin</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0</span></em>, <em class="sig-param"><span class="n">vmax</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">x_tick_rotation</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">45</span></em>, <em class="sig-param"><span class="n">x_tick_ha</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'right'</span></em>, <em class="sig-param"><span class="n">fontsize</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">9</span></em>, <em class="sig-param"><span class="n">z_score_expr</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">z_score_enr</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">grid_color</span><span class="o">=</span><span class="default_value">'grey'</span></em>, <em class="sig-param"><span class="n">grid_lw</span><span class="o">=</span><span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">highlight</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">highlight_lw</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">highlight_lc</span><span class="o">=</span><span class="default_value">'black'</span></em>, <em class="sig-param"><span class="n">figsize</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>Tuple<span class="p">[</span>float<span class="p">, </span>float<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">10, 10</span></em>, <em class="sig-param"><span class="n">use_plotly</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">plotly_height</span><span class="o">=</span><span class="default_value">1000</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/plotting/dotplot.html#dotplot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.plotting.dotplot.dotplot" title="Permalink to this definition"></a></dt>
<dd><p>Function to generate dotplot</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_dotplot</strong> (<em>pd.DataFrame</em>) – A pd.DataFrame with a column Group and varaibles to use for the dotplot</p></li>
<li><p><strong>ax</strong> (<em>plt.axes</em>) – An ax object, only require if use_plotly == False.</p></li>
<li><p><strong>enrichment_variable</strong> (<em>str</em><em>, </em><em>optional</em>) – Variable to use for motif/TF enrichment. Default: ‘Cistrome_AUC;</p></li>
<li><p><strong>region_set_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the column where TF/cistromes are. Default: ‘Cistrome’</p></li>
<li><p><strong>size_var</strong> (<em>str</em><em>, </em><em>optional</em>) – Variable in df_dotplot to code the dot size with. Default: ‘Cistrome_AUC’</p></li>
<li><p><strong>color_var</strong> (<em>str</em><em>, </em><em>optional</em>) – Variable in df_dotplot to code the dot color with. Default: ‘Cistrome_AUC’</p></li>
<li><p><strong>order_group</strong> (<em>list</em><em>, </em><em>optional</em>) – Order to plot the groups by. Default: None (given)</p></li>
<li><p><strong>order_cistromes</strong> (<em>list</em><em>, </em><em>optional</em>) – Order to plot the cistromes by. Default: None (given)</p></li>
<li><p><strong>order_cistromes_by_max</strong> (<em>str</em><em>, </em><em>optional</em>) – Variable to order the cistromes with by values in group, to form a diagonal. Default: None</p></li>
<li><p><strong>cluster</strong> (<em>str</em><em>, </em><em>optional</em>) – Whether to cluster cistromes (or TFs)/groups. It will be overwritten if order_cistromes/order_group/order_cistromes_by_max
are specified.</p></li>
<li><p><strong>n_clust</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of clusters to form if cluster is specified. Default: 2</p></li>
<li><p><strong>min_point_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimal point size. Only available with Cistrome AUC values.</p></li>
<li><p><strong>max_point_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum point size. Only available with Cistrome AUC values.</p></li>
<li><p><strong>cmap</strong> (<em>str</em><em>, </em><em>optional</em>) – Color map to use.</p></li>
<li><p><strong>vmin</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimal color value.</p></li>
<li><p><strong>vmax</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum color value</p></li>
<li><p><strong>x_tick_rotation</strong> (<em>int</em><em>, </em><em>optional</em>) – Rotation of the x-axis label. Only if <cite>use_plotly = False</cite></p></li>
<li><p><strong>fontsize</strong> (<em>int</em><em>, </em><em>optional</em>) – Labels fontsize</p></li>
<li><p><strong>z_score_expr</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to z-normalize expression values</p></li>
<li><p><strong>z_score_enr</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to z-normalize enrichment values</p></li>
<li><p><strong>grid_color</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for grid. Only if <cite>use_plotly = False</cite></p></li>
<li><p><strong>grid_lw</strong> (<em>float</em><em>, </em><em>optional</em>) – Line width of grid. Only if <cite>use_plotly = False</cite></p></li>
<li><p><strong>highlight</strong> (<em>list</em><em>, </em><em>optional</em>) – Whether to highlight specific TFs/cistromes. Only if <cite>use_plotly = False</cite></p></li>
<li><p><strong>highlight_lw</strong> (<em>float</em><em>, </em><em>optional</em>) – Line width of highlight. Only if <cite>use_plotly = False</cite></p></li>
<li><p><strong>highlight_lc</strong> (<em>str</em><em>, </em><em>optional</em>) – Color of highlight. Only if <cite>use_plotly = False</cite></p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Figure size. Only if <cite>use_plotly = False</cite></p></li>
<li><p><strong>use_plotly</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use plotly for plotting. Default: True</p></li>
<li><p><strong>plotly_height</strong> (<em>int</em><em>, </em><em>optional</em>) – Height of the plotly plot. Width will be adjusted accordingly</p></li>
<li><p><strong>save</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to save dotplot as file</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.plotting.dotplot.generate_dotplot_df_AUC">
<code class="sig-prename descclassname">scenicplus.plotting.dotplot.</code><code class="sig-name descname">generate_dotplot_df_AUC</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">auc_key</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">enrichment_key</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">group_variable</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">subset</span><span class="p">:</span> <span class="n">list</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">subset_cistromes</span><span class="p">:</span> <span class="n">list</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">use_pseudobulk</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">normalize_expression</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">standardize_expression</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">standardize_auc</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/plotting/dotplot.html#generate_dotplot_df_AUC"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.plotting.dotplot.generate_dotplot_df_AUC" title="Permalink to this definition"></a></dt>
<dd><p>Function to generate dotplot dataframe from cistrome AUC enrichment</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<cite>class::SCENICPLUS</cite>) – A <code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code> object with motif enrichment results from pycistarget (<cite>scplus_obj.menr</cite>).</p></li>
<li><p><strong>enrichment_key</strong> (<em>str</em>) – Key of the motif enrichment result to use.</p></li>
<li><p><strong>group_variable</strong> (<em>str</em>) – Group variable to use to calculate TF expression per group. Levels of this variable should match with the entries in the
selected motif enrichment dictionary.</p></li>
<li><p><strong>subset</strong> (<em>list</em><em>, </em><em>optional</em>) – Subset of classes to use. Default: None (use all)</p></li>
<li><p><strong>subset_cistromes</strong> (<em>list</em><em>, </em><em>optional</em>) – List of cistromes to use. Default: None (use all)</p></li>
<li><p><strong>use_pseudobulk</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use pseudobulk to calculate gene expression. Only available if there is pseudobulk profiles for group_variable.
Default: False</p></li>
<li><p><strong>normalize_expression</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to log(CPM) normalize gene expression. Default: True</p></li>
<li><p><strong>standardize_expression</strong> (<em>bool</em><em>, </em><em>optional</em>) – Wheter to standarize gene expression between o and 1. Default: False</p></li>
<li><p><strong>standardize_auc</strong> (<em>bool</em><em>, </em><em>optional</em>) – Wheter to standarize AUC values between o and 1. Default: False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>A data frame with enrichment values per TF and cell group to use as input for <cite>dotplot()</cite>.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.plotting.dotplot.generate_dotplot_df_motif_enrichment">
<code class="sig-prename descclassname">scenicplus.plotting.dotplot.</code><code class="sig-name descname">generate_dotplot_df_motif_enrichment</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">enrichment_key</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">group_variable</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">barcode_groups</span><span class="p">:</span> <span class="n">dict</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">subset</span><span class="p">:</span> <span class="n">list</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">subset_TFs</span><span class="p">:</span> <span class="n">list</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">use_pseudobulk</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">use_only_direct</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">normalize_expression</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">standardize_expression</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/plotting/dotplot.html#generate_dotplot_df_motif_enrichment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.plotting.dotplot.generate_dotplot_df_motif_enrichment" title="Permalink to this definition"></a></dt>
<dd><p>Function to generate dotplot dataframe from motif enrichment results</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<cite>class::SCENICPLUS</cite>) – A <code class="xref py py-class docutils literal notranslate"><span class="pre">SCENICPLUS</span></code> object with motif enrichment results from pycistarget (<cite>scplus_obj.menr</cite>).</p></li>
<li><p><strong>enrichment_key</strong> (<em>str</em>) – Key of the motif enrichment result to use.</p></li>
<li><p><strong>group_variable</strong> (<em>str</em><em>, </em><em>optional</em>) – Group variable to use to calculate TF expression per group. Levels of this variable should match with the entries in the
selected motif enrichment dictionary. Only required if barcode groups is not provided.</p></li>
<li><p><strong>barcode_groups</strong> (<em>Dict</em><em>, </em><em>optional</em>) – A dictionary containing cell barcodes per class to calculate TF expression per group. Keys of the dictionary should match
the keys in the selected motif enrichment dictionary. Only required if group_variable is not provided.</p></li>
<li><p><strong>subset</strong> (<em>list</em><em>, </em><em>optional</em>) – Subset of classes to use. Default: None (use all)</p></li>
<li><p><strong>subset_TFs</strong> (<em>list</em><em>, </em><em>optional</em>) – List of TFs to use. Default: None (use all)</p></li>
<li><p><strong>use_pseudobulk</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use pseudobulk to calculate gene expression. Only available if there is pseudobulk profiles for group_variable.
Default: False</p></li>
<li><p><strong>use_only_direct</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to only use NES values of directly annotated motifs. Default: False</p></li>
<li><p><strong>normalize_expression</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to log(CPM) normalize gene expression. Default: True</p></li>
<li><p><strong>standardize_expression</strong> (<em>bool</em><em>, </em><em>optional</em>) – Wheter to standarize gene expression between o and 1. Default: False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>A data frame with enrichment values per TF and cell group to use as input for <cite>dotplot()</cite>.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="export">
<h3>Export<a class="headerlink" href="#export" title="Permalink to this headline"></a></h3>
<div class="section" id="module-scenicplus.loom">
<span id="export-to-loom"></span><h4>Export to loom<a class="headerlink" href="#module-scenicplus.loom" title="Permalink to this headline"></a></h4>
<p>export SCENIC+ object to target genes and target regions loom file.</p>
<p>These files can be visualized in the SCope single cell viewer.</p>
<dl class="py function">
<dt id="scenicplus.loom.export_to_loom">
<code class="sig-prename descclassname">scenicplus.loom.</code><code class="sig-name descname">export_to_loom</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">signature_key</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">out_fname</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">eRegulon_metadata_key</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'eRegulon_metadata'</span></em>, <em class="sig-param"><span class="n">auc_key</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'eRegulon_AUC'</span></em>, <em class="sig-param"><span class="n">auc_thr_key</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'eRegulon_AUC_thresholds'</span></em>, <em class="sig-param"><span class="n">keep_direct_and_extended_if_not_direct</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>bool<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">selected_features</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">selected_cells</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">cluster_annotation</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">tree_structure</span><span class="p">:</span> <span class="n">Sequence<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">()</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">nomenclature</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Unknown'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/loom.html#export_to_loom"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.loom.export_to_loom" title="Permalink to this definition"></a></dt>
<dd><p>Create SCope [Davie et al, 2018] compatible loom files</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scplus_obj</strong> (<em>class::SCENICPLUS</em>) – A SCENIC+ object with eRegulons, AUCs, and AUC thresholds computed</p></li>
<li><p><strong>signature_key</strong> (<em>str</em>) – Whether a ‘Gene_based’ or a ‘Region_based’ file should be produced. Possible values: ‘Gene_based’ or ‘Region_based’</p></li>
<li><p><strong>out_fname</strong> (<em>str</em>) – Path to output file.</p></li>
<li><p><strong>eRegulon_metadata_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Slot where the eRegulon metadata is stored.</p></li>
<li><p><strong>auc_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Slot where the eRegulon AUC are stored</p></li>
<li><p><strong>auc_thr_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Slot where AUC thresholds are stored</p></li>
<li><p><strong>keep_direct_and_extended_if_not_direct</strong> (<em>bool</em><em>, </em><em>optional</em>) – Keep only direct eregulons and add extended ones only if there is not a direct one.</p></li>
<li><p><strong>selected_features</strong> (<em>List</em><em>, </em><em>optional</em>) – A list with selected genes/region to use.</p></li>
<li><p><strong>selected_cells</strong> (<em>List</em><em>, </em><em>optional</em>) – A list with selected cells to use.</p></li>
<li><p><strong>cluster_annotations</strong> (<em>List</em><em>, </em><em>optional</em>) – A list with variables to use as cluster annotations. By default, those included in the DEGs/DARs dictionary will be used (plus those specificed here)</p></li>
<li><p><strong>tree_structure</strong> (<em>sequence</em><em>, </em><em>optional</em>) – A sequence of strings that defines the category tree structure. Needs to be a sequence of strings with three elements. Default: ()</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – The title for this loom file. If None than the basename of the filename is used as the title. Default: None</p></li>
<li><p><strong>nomenclature</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the genome. Default: ‘Unknown’</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>Davie, K., Janssens, J., Koldere, D., De Waegeneer, M., Pech, U., Kreft, Ł., … &amp; Aerts, S. (2018). A single-cell transcriptome atlas of the
aging Drosophila brain. Cell, 174(4), 982-998.
Van de Sande, B., Flerin, C., Davie, K., De Waegeneer, M., Hulselmans, G., Aibar, S., … &amp; Aerts, S. (2020). A scalable SCENIC
workflow for single-cell gene regulatory network analysis. Nature Protocols, 15(7), 2247-2276.</p>
</dd></dl>

</div>
<div class="section" id="module-scenicplus.hdf5">
<span id="export-to-hdf5"></span><h4>Export to hdf5<a class="headerlink" href="#module-scenicplus.hdf5" title="Permalink to this headline"></a></h4>
<p>Write SCENIC+ object to hdf5 file.</p>
<dl class="py function">
<dt id="scenicplus.hdf5.read_h5ad">
<code class="sig-prename descclassname">scenicplus.hdf5.</code><code class="sig-name descname">read_h5ad</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">f_name</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a><a class="reference internal" href="_modules/scenicplus/hdf5.html#read_h5ad"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.hdf5.read_h5ad" title="Permalink to this definition"></a></dt>
<dd><p>Read SCENIC+ object from hdf5 file.</p>
<dl class="simple">
<dt>fname:str</dt><dd><p>Path to hdf5 file.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="scenicplus.hdf5.write_hdf5">
<code class="sig-prename descclassname">scenicplus.hdf5.</code><code class="sig-name descname">write_hdf5</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scplus_obj</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#scenicplus.scenicplus_class.SCENICPLUS" title="scenicplus.scenicplus_class.SCENICPLUS">scenicplus.scenicplus_class.SCENICPLUS</a></span></em>, <em class="sig-param"><span class="n">f_name</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">force_sparse</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/scenicplus/hdf5.html#write_hdf5"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#scenicplus.hdf5.write_hdf5" title="Permalink to this definition"></a></dt>
<dd><p>Write SCENIC+ object to hdf5</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>scplus_obj: SCENICPLUS</strong></dt><dd><p>An instance of scenicplus_class.SCENICPLUS</p>
</dd>
<dt><strong>fname: str</strong></dt><dd><p>Path to out file.</p>
</dd>
<dt><strong>force_sparse: bool = True</strong></dt><dd><p>Wether or not to force sparsity of expression and chromatin accessibility matrix.</p>
</dd>
<dt><strong>verbose = False</strong></dt><dd><p>Wether or not to print debug messages.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
</div>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="install.html" class="btn btn-neutral float-left" title="Install" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="tutorials.html" class="btn btn-neutral float-right" title="Tutorials" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Seppe De Winter, Carmen Bravo Gonzalez Blas.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>